// web/src/pages/contractor/Dashboard.tsx
import React, { useState, useEffect } from 'react';
import { User } from 'shared-types';
import { Building, PlusCircle } from 'lucide-react';
import ActivityTimeline from '../../components/ActivityTimeline';

interface ContractorDashboardProps {
  user: User;
  profileId: string;
  onLogout: () => void;
}

interface Project {
  id: string;
  customer_id: string;
  customer_name: string;
  status: string;
  scheduled_start: string;
  requirement_text: string;
  payment_amount: number | null;
  payment_status: string | null;
}

interface WorkerInfo {
  id: string;
  user_id: string;
  full_name: string;
  skills: string[];
  trust_score: number;
}

export const ContractorDashboard: React.FC<ContractorDashboardProps> = ({ user, profileId, onLogout }) => {
  const [projects, setProjects] = useState<Project[]>([]);
  const [projectAssignments, setProjectAssignments] = useState<{ [projectId: string]: any[] }>({});
  
  // Worker Search states for assignment
  const [assigningProjectId, setAssigningProjectId] = useState<string | null>(null);
  const [searchSkill, setSearchSkill] = useState<string>('mason');
  const [workers, setWorkers] = useState<WorkerInfo[]>([]);
  const [searching, setSearching] = useState<boolean>(false);
  const [assigningWorkerId, setAssigningWorkerId] = useState<string | null>(null);

  // Engagement & Negotiation States
  const [engagements, setEngagements] = useState<any[]>([]);
  const [counterProposalAmount, setCounterProposalAmount] = useState<{ [engagementId: string]: string }>({});

  // Modals / Overlays
  const [payingProject, setPayingProject] = useState<any | null>(null);
  const [payAmount, setPayAmount] = useState<string>('1500');
  const [payMethod, setPayMethod] = useState<'UPI_VERIFIED' | 'CASH_ATTESTED' | 'BANK_VERIFIED'>('UPI_VERIFIED');

  const [ratingProject, setRatingProject] = useState<any | null>(null);
  const [ratingScore, setRatingScore] = useState<number>(5);
  const [ratingComment, setRatingComment] = useState<string>('');

  useEffect(() => {
    fetchProjects();
    fetchEngagements();
  }, [profileId]);

  const fetchProjects = async () => {
    try {
      const res = await fetch(`/api/bookings?userId=${profileId}&role=CONTRACTOR`);
      if (res.ok) {
        const data = await res.json();
        setProjects(data);
        // Fetch assignments for each project
        data.forEach((p: any) => {
          fetchProjectAssignments(p.id);
        });
      }
    } catch (err) {
      console.error(err);
    }
  };

  const fetchProjectAssignments = async (projectId: string) => {
    try {
      const res = await fetch(`/api/bookings/${projectId}/assignments`);
      if (res.ok) {
        const data = await res.json();
        setProjectAssignments(prev => ({ ...prev, [projectId]: data }));
      }
    } catch (err) {
      console.error(err);
    }
  };

  const fetchEngagements = async () => {
    try {
      const res = await fetch(`/api/engagements?userId=${user.id}&role=CONTRACTOR`);
      if (res.ok) {
        const data = await res.json();
        setEngagements(data);
      }
    } catch (e) {
      console.error(e);
    }
  };

  const handleProposePriceOffer = async (engagementId: string) => {
    const amountStr = counterProposalAmount[engagementId];
    if (!amountStr || isNaN(parseFloat(amountStr))) {
      alert('Please enter a valid counter-offer amount.');
      return;
    }

    try {
      const res = await fetch(`/api/engagements/${engagementId}/propose`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          offered_by: user.id,
          amount: parseFloat(amountStr),
          note: 'Contractor Counter Proposal'
        })
      });

      if (!res.ok) throw new Error('Failed to submit counter offer');
      alert('Counter offer proposed successfully!');
      setCounterProposalAmount(prev => ({ ...prev, [engagementId]: '' }));
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleRespondToProposal = async (engagementId: string, response: 'ACCEPTED' | 'REJECTED') => {
    try {
      const res = await fetch(`/api/engagements/${engagementId}/respond`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ response })
      });

      if (!res.ok) throw new Error('Failed to respond to proposal');
      alert(`Proposal ${response.toLowerCase()} successfully!`);
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleUpdateEngagementStatus = async (engagementId: string, status: string) => {
    try {
      const res = await fetch(`/api/engagements/${engagementId}/status`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ status })
      });

      if (!res.ok) throw new Error('Failed to update status');
      alert(`Status updated to ${status}!`);
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleStaffWorker = async (workerId: string) => {
    if (!assigningProjectId) return;
    try {
      const parentEng = engagements.find(e => e.id === assigningProjectId);
      if (!parentEng) {
        alert("Parent engagement not found.");
        return;
      }

      const res = await fetch(`/api/requests/${parentEng.request_id}/engage`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          mode: 'DIRECT_WORKER',
          initiator_id: user.id,
          counterparty_id: workerId,
          parent_engagement_id: parentEng.id,
          initial_amount: 500,
          note: `Staff assignment by Contractor ${user.full_name}`
        })
      });

      if (!res.ok) {
        const errData = await res.json();
        throw new Error(errData.error || 'Failed to staff worker');
      }

      alert('Worker staffing invitation sent successfully!');
      setWorkers([]);
      setAssigningProjectId(null);
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleAcceptProject = async (projectId: string) => {
    try {
      const res = await fetch(`/api/bookings/${projectId}/accept`, { method: 'POST' });
      if (!res.ok) throw new Error('Failed to accept project');
      alert('Project accepted successfully!');
      fetchProjects();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleStartProject = async (projectId: string) => {
    try {
      const res = await fetch(`/api/bookings/${projectId}/start`, { method: 'POST' });
      if (!res.ok) throw new Error('Failed to start project');
      alert('Project is now IN_PROGRESS!');
      fetchProjects();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleCompleteProject = async (projectId: string) => {
    try {
      const res = await fetch(`/api/bookings/${projectId}/complete`, { method: 'POST' });
      if (!res.ok) throw new Error('Failed to complete project');
      alert('Project marked as COMPLETED!');
      fetchProjects();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleSearchWorkers = async () => {
    if (!searchSkill.trim()) return;
    setSearching(true);
    try {
      const res = await fetch(`/api/customer/find-workers?skill=${encodeURIComponent(searchSkill.toLowerCase())}`);
      if (res.ok) {
        const data = await res.json();
        setWorkers(data);
      }
    } catch (err) {
      console.error(err);
    } finally {
      setSearching(false);
    }
  };

  const handleAssignWorker = async (workerId: string) => {
    if (!assigningProjectId) return;
    setAssigningWorkerId(workerId);
    try {
      const res = await fetch('/api/assignments', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          booking_id: assigningProjectId,
          worker_id: workerId,
          contractor_id: profileId,
          remarks: `Assigned to project by Contractor ${user.full_name}`
        })
      });

      if (!res.ok) {
        const errData = await res.json();
        throw new Error(errData.error || 'Failed to assign worker');
      }

      alert('Worker assigned successfully! Worker notified.');
      setWorkers([]);
      setAssigningProjectId(null);
      fetchProjectAssignments(assigningProjectId);
    } catch (err: any) {
      alert(err.message);
    } finally {
      setAssigningWorkerId(null);
    }
  };

  const handlePaymentSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!payingProject || !payAmount) return;

    try {
      const res = await fetch(`/api/bookings/${payingProject.id}/payments`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          amount: parseFloat(payAmount),
          confirmation_method: payMethod
        })
      });

      if (!res.ok) throw new Error('Payment failed');
      alert('Payment made to worker successfully!');
      setPayingProject(null);
      fetchProjects();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleRatingSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!ratingProject || !ratingScore) return;

    // Use first accepted worker or similar for rating
    const asnList = projectAssignments[ratingProject.id] || [];
    const acceptedAsn = asnList.find(a => a.status === 'ACCEPTED');
    if (!acceptedAsn) {
      alert('No accepted worker on this project to rate!');
      return;
    }

    try {
      const res = await fetch(`/api/bookings/${ratingProject.id}/ratings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          rater_id: user.id,
          ratee_id: acceptedAsn.worker_id,
          score: ratingScore,
          comment: ratingComment
        })
      });

      if (!res.ok) throw new Error('Rating failed');
      alert('Rating submitted successfully!');
      setRatingProject(null);
      setRatingComment('');
      fetchProjects();
    } catch (err: any) {
      alert(err.message);
    }
  };

  return (
    <div style={{ fontFamily: 'Outfit, sans-serif', backgroundColor: '#f1f5f9', minHeight: '100vh', padding: '24px' }}>
      {/* Header Bar */}
      <header style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', backgroundColor: '#ffffff', padding: '16px 24px', border: '1px solid #e2e8f0', borderRadius: '12px', marginBottom: '24px', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
        <div style={{ display: 'flex', alignItems: 'center', gap: '12px' }}>
          <Building size={28} color="#0f766e" />
          <div>
            <h1 style={{ margin: 0, fontSize: '20px', color: '#0f766e', fontWeight: 'bold' }}>Contractor Hub (नियोजक)</h1>
            <p style={{ margin: '2px 0 0 0', fontSize: '13px', color: '#64748b' }}>
              Logged in: <strong style={{ color: '#1e293b' }}>{user.full_name}</strong> | UID: {profileId.slice(0, 8)}
            </p>
          </div>
        </div>
        <button 
          onClick={onLogout} 
          style={{ backgroundColor: '#ffffff', color: '#ef4444', border: '1px solid #fecaca', padding: '8px 16px', fontSize: '13px', fontWeight: '600', borderRadius: '8px', cursor: 'pointer' }}
        >
          Logout Session
        </button>
      </header>

      {/* Main Grid */}
      <div style={{ display: 'grid', gridTemplateColumns: '2.5fr 1.2fr', gap: '24px' }}>
        
        {/* Left Column: Projects */}
        <div>
          {/* Active Contracts & Staffing Panel */}
          <section className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px' }}>
            <h3 style={{ margin: '0 0 16px 0', fontSize: '16px', color: '#1e293b', fontWeight: '700', display: 'flex', alignItems: 'center', gap: '8px' }}>
              🤝 ACTIVE CONTRACTS & STAFFING
            </h3>

            {engagements.filter(e => !e.parent_engagement_id).length === 0 ? (
              <div style={{ padding: '40px', textAlign: 'center', border: '2px dashed #cbd5e1', borderRadius: '8px', color: '#64748b', fontSize: '14px' }}>
                No active parent engagements or negotiations found.
              </div>
            ) : (
              <div style={{ display: 'flex', flexDirection: 'column', gap: '16px' }}>
                {engagements.filter(e => !e.parent_engagement_id).map((eng) => {
                  const lastOffer = eng.offers && eng.offers[0];
                  const isPendingOrNegotiating = eng.status === 'PENDING' || eng.status === 'NEGOTIATING';
                  const lastOfferedByMe = lastOffer && lastOffer.offered_by === user.id;

                  // Find sub-engagements for this project
                  const subs = engagements.filter(sub => sub.parent_engagement_id === eng.id);

                  return (
                    <div key={eng.id} style={{ padding: '16px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#f8fafc' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: '8px' }}>
                        <div>
                          <span style={{ fontSize: '10px', color: '#94a3b8', fontWeight: '500' }}>ENGAGEMENT ID: {eng.id.slice(0, 8)}...</span>
                          <h4 style={{ margin: '4px 0 0 0', fontSize: '14px', fontWeight: '700', color: '#1e293b' }}>
                            Client: {eng.initiator_name} (VIA_CONTRACTOR Mode)
                          </h4>
                          <p style={{ margin: '6px 0', fontSize: '12px', color: '#475569' }}>
                            Scope of Work: <strong>"{eng.request_text}"</strong>
                          </p>
                        </div>
                        <span style={{
                          fontSize: '11px',
                          fontWeight: 'bold',
                          padding: '4px 8px',
                          borderRadius: '6px',
                          backgroundColor: eng.status === 'COMPLETED' ? '#e2f5f3' : (eng.status === 'IN_PROGRESS' ? '#dbeafe' : (eng.status === 'DISPUTED' ? '#fee2e2' : '#f1f5f9')),
                          color: eng.status === 'COMPLETED' ? '#0f766e' : (eng.status === 'IN_PROGRESS' ? '#2563eb' : (eng.status === 'DISPUTED' ? '#dc2626' : '#475569'))
                        }}>
                          {eng.status}
                        </span>
                      </div>

                      {/* Offers Log */}
                      {eng.offers && eng.offers.length > 0 && (
                        <div style={{ marginTop: '10px', padding: '10px', backgroundColor: '#ffffff', borderRadius: '6px', border: '1px solid #e2e8f0' }}>
                          <span style={{ fontSize: '11px', fontWeight: 'bold', color: '#64748b', display: 'block', marginBottom: '4px' }}>Negotiation History:</span>
                          <div style={{ display: 'flex', flexDirection: 'column', gap: '4px' }}>
                            {eng.offers.slice(0, 3).map((o: any) => (
                              <div key={o.id} style={{ display: 'flex', justifyContent: 'space-between', fontSize: '12px', color: '#475569' }}>
                                <span>₹{o.amount} proposed by {o.offered_by_name}</span>
                                <span style={{ fontSize: '10px', color: '#94a3b8' }}>{new Date(o.created_at).toLocaleTimeString()}</span>
                              </div>
                            ))}
                          </div>
                        </div>
                      )}

                      {/* Sub-Engagements (Staffed Workers) */}
                      <div style={{ marginTop: '12px', padding: '12px', backgroundColor: '#eff6ff', borderRadius: '8px', border: '1px solid #bfdbfe' }}>
                        <span style={{ fontSize: '11px', fontWeight: '700', color: '#1e40af', display: 'block', marginBottom: '6px' }}>
                          Staffed Workers (sub-contracts):
                        </span>
                        {subs.length === 0 ? (
                          <div style={{ fontSize: '12px', color: '#60a5fa', fontStyle: 'italic' }}>
                            No workers staffed yet.
                          </div>
                        ) : (
                          <div style={{ display: 'flex', flexDirection: 'column', gap: '6px' }}>
                            {subs.map(sub => (
                              <div key={sub.id} style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', fontSize: '12px', backgroundColor: '#ffffff', padding: '6px 10px', borderRadius: '6px', border: '1px solid #dbeafe' }}>
                                <span style={{ fontWeight: '600', color: '#1e293b' }}>👤 {sub.counterparty_name}</span>
                                <span style={{ fontSize: '11px', fontWeight: 'bold', color: sub.status === 'COMPLETED' ? '#10b981' : '#d97706' }}>
                                  {sub.status} (₹{sub.offers?.[0]?.amount || 500})
                                </span>
                              </div>
                            ))}
                          </div>
                        )}
                      </div>

                      {/* Controls */}
                      <div style={{ marginTop: '12px', display: 'flex', flexWrap: 'wrap', gap: '8px', alignItems: 'center' }}>
                        {isPendingOrNegotiating && (
                          <>
                            {!lastOfferedByMe ? (
                              <>
                                <button
                                  onClick={() => handleRespondToProposal(eng.id, 'ACCEPTED')}
                                  style={{ backgroundColor: '#10b981', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                >
                                  Accept Offer (₹{lastOffer?.amount || 1500})
                                </button>
                                <button
                                  onClick={() => handleRespondToProposal(eng.id, 'REJECTED')}
                                  style={{ backgroundColor: '#ef4444', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                >
                                  Reject Offer
                                </button>
                              </>
                            ) : (
                              <span style={{ fontSize: '12px', color: '#64748b', fontStyle: 'italic' }}>Awaiting client response...</span>
                            )}

                            <div style={{ display: 'flex', gap: '4px', marginLeft: 'auto' }}>
                              <input
                                type="number"
                                placeholder="Counter offer price"
                                value={counterProposalAmount[eng.id] || ''}
                                onChange={(e) => setCounterProposalAmount({ ...counterProposalAmount, [eng.id]: e.target.value })}
                                style={{ width: '130px', padding: '5px 8px', fontSize: '12px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                              />
                              <button
                                onClick={() => handleProposePriceOffer(eng.id)}
                                style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                              >
                                Counter
                              </button>
                            </div>
                          </>
                        )}

                        {(eng.status === 'ACCEPTED' || eng.status === 'IN_PROGRESS') && (
                          <>
                            <button
                              onClick={() => {
                                setAssigningProjectId(eng.id);
                                setSearchSkill('mason');
                                setWorkers([]);
                              }}
                              style={{ backgroundColor: '#2563eb', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                            >
                              ➕ Staff Worker (Assign Sub-Contract)
                            </button>

                            {eng.status === 'ACCEPTED' && (
                              <button
                                onClick={() => handleUpdateEngagementStatus(eng.id, 'IN_PROGRESS')}
                                style={{ backgroundColor: '#10b981', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                              >
                                Start Work
                              </button>
                            )}

                            {eng.status === 'IN_PROGRESS' && (
                              <button
                                onClick={() => handleUpdateEngagementStatus(eng.id, 'COMPLETED')}
                                style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                              >
                                Mark Completed
                              </button>
                            )}
                          </>
                        )}
                      </div>
                    </div>
                  );
                })}
              </div>
            )}
          </section>

          {/* Worker staffing sub-panel */}
          {assigningProjectId && (
            <section className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #bfdbfe', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px', borderLeft: '4px solid #2563eb' }}>
              <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '16px' }}>
                <h4 style={{ margin: 0, fontSize: '15px', color: '#1e3a8a', fontWeight: 'bold' }}>
                  🔍 Staff Worker for Project {assigningProjectId.slice(0, 8)}...
                </h4>
                <button 
                  onClick={() => setAssigningProjectId(null)}
                  style={{ background: 'none', border: 'none', color: '#ef4444', fontSize: '12px', cursor: 'pointer', fontWeight: 'bold' }}
                >
                  Cancel
                </button>
              </div>

              <div style={{ display: 'flex', gap: '8px', marginBottom: '16px' }}>
                <select 
                  className="form-control" 
                  value={searchSkill} 
                  onChange={(e) => setSearchSkill(e.target.value)} 
                  style={{ padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1', flex: 1 }}
                >
                  <option value="mason">Mason (राजमिस्त्री)</option>
                  <option value="plumber">Plumber (नलसाज)</option>
                  <option value="electrician">Electrician (बिजली मिस्त्री)</option>
                  <option value="painter">Painter (रंगसाज)</option>
                  <option value="carpenter">Carpenter (बढ़ई)</option>
                  <option value="laborer">General Helper (मजदूर)</option>
                </select>
                <button 
                  className="btn btn-primary" 
                  onClick={handleSearchWorkers} 
                  disabled={searching}
                  style={{ padding: '8px 16px', backgroundColor: '#2563eb', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer', fontWeight: 'bold' }}
                >
                  {searching ? 'Searching...' : 'Search'}
                </button>
              </div>

              {workers.length === 0 ? (
                <div style={{ padding: '16px', textAlign: 'center', fontSize: '12px', color: '#64748b' }}>
                  No available workers found for this skill category.
                </div>
              ) : (
                <div style={{ display: 'flex', flexDirection: 'column', gap: '10px' }}>
                  {workers.map(w => (
                    <div key={w.id} style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', padding: '10px', border: '1px solid #e2e8f0', borderRadius: '6px' }}>
                      <div>
                        <span style={{ fontWeight: 'bold', fontSize: '13px', color: '#1e293b' }}>{w.full_name}</span>
                        <div style={{ fontSize: '11px', color: '#64748b' }}>Trust Score: {w.trust_score}% | Skills: {w.skills.join(', ')}</div>
                      </div>
                      <button
                        onClick={() => handleStaffWorker(w.user_id)}
                        style={{ backgroundColor: '#10b981', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '11px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                      >
                        Send Staff Invitation
                      </button>
                    </div>
                  ))}
                </div>
              )}
            </section>
          )}

          <section className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
            <h3 style={{ margin: '0 0 16px 0', fontSize: '16px', color: '#1e293b', fontWeight: '700', borderBottom: '1px solid #f1f5f9', paddingBottom: '12px' }}>
              CUSTOMER PROJECT BOOKINGS
            </h3>

            {projects.length === 0 ? (
              <div style={{ padding: '40px', textAlign: 'center', border: '2px dashed #cbd5e1', borderRadius: '8px', color: '#64748b' }}>
                No project bookings received yet.
              </div>
            ) : (
              <div style={{ display: 'flex', flexDirection: 'column', gap: '16px' }}>
                {projects.map((p) => {
                  const asnList = projectAssignments[p.id] || [];
                  return (
                    <div key={p.id} style={{ padding: '18px', border: '1px solid #e2e8f0', borderRadius: '10px', backgroundColor: '#ffffff' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: '10px' }}>
                        <div>
                          <span style={{ fontSize: '11px', color: '#94a3b8', fontWeight: '500' }}>Project ID: {p.id.slice(0, 8)}...</span>
                          <h4 style={{ margin: '4px 0 0 0', fontSize: '15px', fontWeight: '700', color: '#0f766e' }}>
                            Customer: {p.customer_name}
                          </h4>
                        </div>
                        <span style={{ 
                          fontSize: '11px', 
                          fontWeight: 'bold', 
                          padding: '4px 8px', 
                          borderRadius: '6px', 
                          backgroundColor: p.status === 'COMPLETED' ? '#e2f5f3' : (p.status === 'ACCEPTED' || p.status === 'IN_PROGRESS' ? '#fffbeb' : '#f1f5f9'), 
                          color: p.status === 'COMPLETED' ? '#0f766e' : (p.status === 'ACCEPTED' || p.status === 'IN_PROGRESS' ? '#d97706' : '#475569') 
                        }}>
                          {p.status}
                        </span>
                      </div>

                      <p style={{ margin: '8px 0', fontSize: '13px', color: '#475569', backgroundColor: '#f8fafc', padding: '10px', borderRadius: '6px', borderLeft: '3px solid #0f766e' }}>
                        Requirement: <strong>"{p.requirement_text}"</strong>
                      </p>

                      {/* Assignments list */}
                      <div style={{ marginTop: '14px', borderTop: '1px solid #f1f5f9', paddingTop: '12px' }}>
                        <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '8px' }}>
                          <span style={{ fontSize: '12px', fontWeight: '700', color: '#475569' }}>👥 Assigned Workers:</span>
                          {p.status === 'ACCEPTED' && (
                            <button 
                              onClick={() => { setAssigningProjectId(p.id); setWorkers([]); }} 
                              style={{ display: 'flex', alignItems: 'center', gap: '4px', border: 'none', background: 'none', color: '#0f766e', fontSize: '12px', fontWeight: 'bold', cursor: 'pointer' }}
                            >
                              <PlusCircle size={14} /> Assign Worker
                            </button>
                          )}
                        </div>

                        {asnList.length === 0 ? (
                          <div style={{ fontSize: '12px', color: '#94a3b8', padding: '6px 0' }}>No workers assigned to this project yet.</div>
                        ) : (
                          <div style={{ display: 'flex', flexDirection: 'column', gap: '6px' }}>
                            {asnList.map((asn: any) => (
                              <div key={asn.id} style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', fontSize: '12px', padding: '6px 10px', backgroundColor: '#f1f5f9', borderRadius: '6px' }}>
                                <span style={{ fontWeight: '500' }}>👤 {asn.worker_name} ({asn.skills ? JSON.parse(asn.skills).join(', ') : ''})</span>
                                <span style={{ 
                                  fontSize: '10px', 
                                  fontWeight: 'bold',
                                  padding: '2px 6px',
                                  borderRadius: '4px',
                                  backgroundColor: asn.status === 'ACCEPTED' ? '#d1fae5' : asn.status === 'ASSIGNED' ? '#fef3c7' : '#fee2e2',
                                  color: asn.status === 'ACCEPTED' ? '#065f46' : asn.status === 'ASSIGNED' ? '#92400e' : '#991b1b'
                                }}>
                                  {asn.status}
                                </span>
                              </div>
                            ))}
                          </div>
                        )}
                      </div>

                      {/* Project Control Actions */}
                      <div style={{ display: 'flex', gap: '8px', justifyContent: 'flex-end', marginTop: '16px', borderTop: '1px solid #f1f5f9', paddingTop: '12px' }}>
                        {p.status === 'REQUESTED' && (
                          <button 
                            onClick={() => handleAcceptProject(p.id)} 
                            style={{ ...actionBtnStyle, backgroundColor: '#0f766e', color: '#ffffff', border: 'none' }}
                          >
                            Accept Project
                          </button>
                        )}
                        {p.status === 'ACCEPTED' && (
                          <button 
                            onClick={() => handleStartProject(p.id)} 
                            style={{ ...actionBtnStyle, backgroundColor: '#d97706', color: '#ffffff', border: 'none' }}
                          >
                            Start Project Work
                          </button>
                        )}
                        {p.status === 'IN_PROGRESS' && (
                          <button 
                            onClick={() => handleCompleteProject(p.id)} 
                            style={{ ...actionBtnStyle, backgroundColor: '#059669', color: '#ffffff', border: 'none' }}
                          >
                            Complete Project
                          </button>
                        )}
                        {p.status === 'COMPLETED' && (
                          <>
                            <button 
                              onClick={() => setPayingProject(p)} 
                              style={{ ...actionBtnStyle, backgroundColor: '#2563eb', color: '#ffffff', border: 'none' }}
                            >
                              💸 Pay Worker
                            </button>
                            <button 
                              onClick={() => setRatingProject(p)} 
                              style={actionBtnStyle}
                            >
                              ⭐ Rate Worker
                            </button>
                          </>
                        )}
                      </div>
                    </div>
                  );
                })}
              </div>
            )}
          </section>
        </div>

        {/* Right Column: Timeline & Assignment Panel */}
        <div>
          {assigningProjectId && (
            <section className="card" style={{ backgroundColor: '#ffffff', padding: '20px', borderRadius: '12px', border: '1px solid #99f6e4', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px' }}>
              <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '12px' }}>
                <h3 style={{ margin: 0, fontSize: '14px', color: '#0f766e', fontWeight: 'bold' }}>Assign Worker to Project</h3>
                <button onClick={() => setAssigningProjectId(null)} style={{ background: 'none', border: 'none', color: '#64748b', fontSize: '12px', cursor: 'pointer' }}>Close</button>
              </div>
              <div className="form-group" style={{ marginBottom: '12px' }}>
                <label className="form-label">Search Worker Skill</label>
                <div style={{ display: 'flex', gap: '8px' }}>
                  <input 
                    type="text" 
                    className="form-control" 
                    value={searchSkill} 
                    onChange={(e) => setSearchSkill(e.target.value)} 
                    style={{ flex: 1, padding: '6px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                  />
                  <button 
                    onClick={handleSearchWorkers} 
                    style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', borderRadius: '6px', cursor: 'pointer', fontWeight: 'bold' }}
                    disabled={searching}
                  >
                    Search
                  </button>
                </div>
              </div>

              {workers.length > 0 ? (
                <div style={{ display: 'flex', flexDirection: 'column', gap: '8px', marginTop: '12px' }}>
                  {workers.map(w => (
                    <div key={w.id} style={{ padding: '10px', border: '1px solid #e2e8f0', borderRadius: '6px', backgroundColor: '#f8fafc' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
                        <div>
                          <strong style={{ fontSize: '13px' }}>{w.full_name}</strong>
                          <div style={{ fontSize: '10px', color: '#64748b' }}>Skills: {w.skills ? w.skills.join(', ') : ''}</div>
                        </div>
                        <button 
                          onClick={() => handleAssignWorker(w.id)}
                          style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '4px 8px', fontSize: '11px', borderRadius: '4px', cursor: 'pointer' }}
                          disabled={assigningWorkerId === w.id}
                        >
                          {assigningWorkerId === w.id ? 'Assigning...' : 'Assign'}
                        </button>
                      </div>
                    </div>
                  ))}
                </div>
              ) : (
                <div style={{ fontSize: '11px', color: '#94a3b8', textAlign: 'center', padding: '12px 0' }}>Search and select a worker above.</div>
              )}
            </section>
          )}

          <section className="card" style={{ backgroundColor: '#ffffff', padding: '20px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
            <ActivityTimeline userId={user.id} />
          </section>
        </div>
      </div>

      {/* Pay Worker Modal */}
      {payingProject && (
        <div style={backdropStyle}>
          <div style={modalStyle}>
            <div style={{ padding: '20px', borderBottom: '1px solid #cbd5e1' }}>
              <h3 style={{ margin: 0, fontSize: '15px', color: '#1e293b', fontWeight: 'bold' }}>Pay Worker</h3>
            </div>
            <form onSubmit={handlePaymentSubmit}>
              <div style={{ padding: '20px' }}>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label">Payment Amount (Rs.)</label>
                  <input 
                    type="number"
                    className="form-control"
                    value={payAmount}
                    onChange={(e) => setPayAmount(e.target.value)}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                    required
                  />
                </div>
                <div className="form-group">
                  <label className="form-label">Payment Channel</label>
                  <select 
                    className="form-control"
                    value={payMethod}
                    onChange={(e) => setPayMethod(e.target.value as any)}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                  >
                    <option value="UPI_VERIFIED">UPI Transaction</option>
                    <option value="CASH_ATTESTED">Cash Hand-to-Hand</option>
                    <option value="BANK_VERIFIED">Bank Transfer</option>
                  </select>
                </div>
              </div>
              <div style={modalFooterStyle}>
                <button type="button" onClick={() => setPayingProject(null)} style={cancelBtnStyle}>Cancel</button>
                <button type="submit" className="btn btn-primary" style={{ padding: '6px 12px', backgroundColor: '#0f766e', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer' }}>attest Payment</button>
              </div>
            </form>
          </div>
        </div>
      )}

      {/* Rate Worker Modal */}
      {ratingProject && (
        <div style={backdropStyle}>
          <div style={modalStyle}>
            <div style={{ padding: '20px', borderBottom: '1px solid #cbd5e1' }}>
              <h3 style={{ margin: 0, fontSize: '15px', color: '#1e293b', fontWeight: 'bold' }}>Rate Worker Performance</h3>
            </div>
            <form onSubmit={handleRatingSubmit}>
              <div style={{ padding: '20px' }}>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label">Score (1 to 5 Stars)</label>
                  <select 
                    className="form-control"
                    value={ratingScore} 
                    onChange={(e) => setRatingScore(parseInt(e.target.value))}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                  >
                    <option value={5}>5 Stars - Excellent</option>
                    <option value={4}>4 Stars - Good</option>
                    <option value={3}>3 Stars - Average</option>
                    <option value={2}>2 Stars - Poor</option>
                    <option value={1}>1 Star - Very Poor</option>
                  </select>
                </div>
                <div className="form-group">
                  <label className="form-label">Comments</label>
                  <textarea 
                    className="form-control"
                    value={ratingComment}
                    onChange={(e) => setRatingComment(e.target.value)}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1', minHeight: '60px' }}
                    required
                  />
                </div>
              </div>
              <div style={modalFooterStyle}>
                <button type="button" onClick={() => setRatingProject(null)} style={cancelBtnStyle}>Cancel</button>
                <button type="submit" className="btn btn-primary" style={{ padding: '6px 12px', backgroundColor: '#0f766e', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer' }}>Submit Rating</button>
              </div>
            </form>
          </div>
        </div>
      )}

    </div>
  );
};

const actionBtnStyle: React.CSSProperties = {
  backgroundColor: '#f1f5f9',
  border: '1px solid #cbd5e1',
  color: '#334155',
  padding: '6px 12px',
  fontSize: '12px',
  fontWeight: 'bold',
  borderRadius: '6px',
  cursor: 'pointer'
};

const backdropStyle: React.CSSProperties = {
  position: 'fixed',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  backgroundColor: 'rgba(15, 23, 42, 0.6)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  zIndex: 999
};

const modalStyle: React.CSSProperties = {
  backgroundColor: '#ffffff',
  width: '100%',
  maxWidth: '440px',
  borderRadius: '8px',
  boxShadow: '0 10px 15px -3px rgb(0 0 0 / 0.1)',
  overflow: 'hidden'
};

const modalFooterStyle: React.CSSProperties = {
  padding: '12px 20px',
  backgroundColor: '#f8fafc',
  borderTop: '1px solid #cbd5e1',
  display: 'flex',
  justifyContent: 'flex-end',
  gap: '8px'
};

const cancelBtnStyle: React.CSSProperties = {
  background: 'none',
  border: '1px solid #cbd5e1',
  padding: '6px 12px',
  borderRadius: '6px',
  fontSize: '13px',
  cursor: 'pointer',
  color: '#475569'
};

export default ContractorDashboard;
