// web/src/pages/Login.tsx
import React, { useState, useEffect } from 'react';
import { 
  Shield, 
  Lock, 
  Phone, 
  HardHat, 
  Home, 
  Building2, 
  Users, 
  CheckCircle2, 
  Briefcase, 
  Fingerprint, 
  UserCheck, 
  AlertTriangle, 
  CreditCard,
  Eye,
  EyeOff,
  Star
} from 'lucide-react';
import './Login.css';

interface LoginProps {
  onLoginSuccess: (user: any, profileId: string) => void;
}

export const Login: React.FC<LoginProps> = ({ onLoginSuccess }) => {
  const API_URL = import.meta.env.VITE_API_URL || '';
  const [role, setRole] = useState<'WORKER' | 'CUSTOMER' | 'CONTRACTOR' | 'ADMIN'>('WORKER');
  const [isRegister, setIsRegister] = useState<boolean>(false);
  const [showPassword, setShowPassword] = useState<boolean>(false);
  const [rememberMe, setRememberMe] = useState<boolean>(false);
  
  // Input fields
  const [phone, setPhone] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [fullName, setFullName] = useState<string>('');
  const [email, setEmail] = useState<string>('');
  const [companyName, setCompanyName] = useState<string>('');
  const [homeAddress, setHomeAddress] = useState<string>('');
  const [preferredLanguage, setPreferredLanguage] = useState<string>('en');

  // Dynamic Skill Taxonomy Onboarding
  const [skillsTaxonomy, setSkillsTaxonomy] = useState<{ id: string; category: string; name: string }[]>([]);
  const [selectedSkills, setSelectedSkills] = useState<string[]>([]);

  const [loading, setLoading] = useState<boolean>(false);
  const [errorText, setErrorText] = useState<string | null>(null);

  useEffect(() => {
    document.body.classList.add('login-page-active');
    document.documentElement.classList.add('login-page-active');

    fetchSkills();
    
    // Check initial path route
    const path = window.location.pathname;
    if (path.includes('/auth/customer')) {
      setRole('CUSTOMER');
    } else if (path.includes('/auth/contractor')) {
      setRole('CONTRACTOR');
    } else if (path.includes('/auth/admin')) {
      setRole('ADMIN');
      setIsRegister(false);
    } else {
      setRole('WORKER');
    }

    return () => {
      document.body.classList.remove('login-page-active');
      document.documentElement.classList.remove('login-page-active');
    };
  }, []);

  const fetchSkills = async () => {
    try {
      const res = await fetch(`${API_URL}/api/customer/skills`);
      if (res.ok) {
        const data = await res.json();
        setSkillsTaxonomy(data);
      }
    } catch (e) {
      console.error('Failed to load skills taxonomy', e);
    }
  };

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    setErrorText(null);
    setLoading(true);

    const url = isRegister
      ? `${API_URL}/api/auth/register`
      : `${API_URL}/api/auth/login`;

    const payload = isRegister ? {
      phone,
      password,
      full_name: fullName,
      role,
      email: email || undefined,
      company_name: role === 'CONTRACTOR' ? companyName : undefined,
      skills: role === 'WORKER' ? selectedSkills : undefined,
      home_address: role === 'CUSTOMER' ? homeAddress : undefined,
      preferred_language: role === 'CUSTOMER' ? preferredLanguage : undefined,
      lat: role === 'CUSTOMER' ? 19.1197 : 28.6139, // Default coordinates
      lng: role === 'CUSTOMER' ? 72.8464 : 77.2090,
      device_fingerprint: 'win_chrome_pc_901',
      ip_address: '127.0.0.1'
    } : {
      phone,
      password,
      device_fingerprint: 'win_chrome_pc_901',
      ip_address: '127.0.0.1'
    };

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const data = await response.json();
      if (!response.ok) throw new Error(data.error || 'Authentication failed');

      // Fetch profile info
      const meRes = await fetch(`${API_URL}/api/auth/me`);
      if (meRes.ok) {
        const meData = await meRes.json();
        onLoginSuccess(meData.user, meData.profileId);
      } else {
        onLoginSuccess(data.user, '');
      }

    } catch (err: any) {
      setErrorText(err.message);
    } finally {
      setLoading(false);
    }
  };

  // Group skills by category
  const groupedSkills = skillsTaxonomy.reduce((acc, curr) => {
    if (!acc[curr.category]) {
      acc[curr.category] = [];
    }
    acc[curr.category].push(curr.name);
    return acc;
  }, {} as Record<string, string[]>);

  const getRoleIcon = (r: string) => {
    switch(r) {
      case 'WORKER': return <HardHat size={14} />;
      case 'CUSTOMER': return <Home size={14} />;
      case 'CONTRACTOR': return <Building2 size={14} />;
      case 'ADMIN': return <Shield size={14} />;
      default: return <Users size={14} />;
    }
  };

  const getRoleLabel = (r: string) => {
    switch(r) {
      case 'WORKER': return 'Worker';
      case 'CUSTOMER': return 'Customer';
      case 'CONTRACTOR': return 'Contractor';
      case 'ADMIN': return 'Admin';
      default: return r;
    }
  };

  return (
    <div className="login-layout">
      {/* Left Side - Hero Area */}
      <div className="login-hero">
        <div className="hero-header-container">
          <div className="hero-branding">
            <div className="branding-top">
              <Shield className="hero-logo-icon" size={28} />
              <h1 className="hero-logo-text">LABOUR<span>LINK</span></h1>
            </div>
            <span className="hero-logo-subtitle">Trust Infra</span>
          </div>

          <div className="trust-badge">
            <CheckCircle2 size={16} />
            <span>Verified. Trusted. Transparent.</span>
          </div>

          <h2 className="hero-headline">India's Trusted<br/><span>Labour</span> Network</h2>
          <p className="hero-description">
            Connecting verified workers, reliable contractors and customers through an AI-powered trust infrastructure.
          </p>

          <div className="stats-container">
            <div className="stat-card">
              <div className="stat-header">
                <div className="stat-icon-wrapper">
                  <Users size={16} />
                </div>
                <span className="stat-value">10,000+</span>
              </div>
              <span className="stat-label">Verified Workers</span>
            </div>
            <div className="stat-card">
              <div className="stat-header">
                <div className="stat-icon-wrapper">
                  <Shield size={16} />
                </div>
                <span className="stat-value">98%</span>
              </div>
              <span className="stat-label">Trust Accuracy</span>
            </div>
            <div className="stat-card">
              <div className="stat-header">
                <div className="stat-icon-wrapper">
                  <Briefcase size={16} />
                </div>
                <span className="stat-value">5,000+</span>
              </div>
              <span className="stat-label">Jobs Completed</span>
            </div>
          </div>
        </div>

        {/* Central Worker Group Image */}
        <div className="hero-image-wrapper">
          <img src="/hero.png" alt="Diverse Indian Workforce" className="hero-image" />
        </div>

        {/* Features Strip */}
        <div className="trust-features-strip">
          <div className="trust-feature-item">
            <div className="trust-feature-icon-wrapper">
              <Fingerprint size={16} />
            </div>
            <div className="trust-feature-info">
              <span className="trust-feature-title">Aadhaar</span>
              <span className="trust-feature-desc">Verified</span>
            </div>
          </div>

          <div className="trust-feature-item">
            <div className="trust-feature-icon-wrapper">
              <UserCheck size={16} />
            </div>
            <div className="trust-feature-info">
              <span className="trust-feature-title">Background</span>
              <span className="trust-feature-desc">Verification</span>
            </div>
          </div>

          <div className="trust-feature-item">
            <div className="trust-feature-icon-wrapper">
              <AlertTriangle size={16} />
            </div>
            <div className="trust-feature-info">
              <span className="trust-feature-title">Fraud Detection</span>
              <span className="trust-feature-desc">Engine</span>
            </div>
          </div>

          <div className="trust-feature-item">
            <div className="trust-feature-icon-wrapper">
              <CreditCard size={16} />
            </div>
            <div className="trust-feature-info">
              <span className="trust-feature-title">Secure</span>
              <span className="trust-feature-desc">Payments</span>
            </div>
          </div>
        </div>
      </div>

      {/* Right Side - Auth Area */}
      <div className="login-auth-container">
        <div className="auth-card">
          <div className="auth-header">
            <div className="auth-header-icon-container">
              <Shield size={22} />
            </div>
            <h2 className="auth-title">Welcome Back!</h2>
            <p className="auth-subtitle">Login to your LabourLink account</p>
          </div>

          <div className="role-cards">
            {(['WORKER', 'CUSTOMER', 'CONTRACTOR', 'ADMIN'] as const).map((r) => {
              const isActive = role === r;
              const title = getRoleLabel(r);
              return (
                <div 
                  key={r}
                  className={`role-card ${isActive ? 'active' : ''}`}
                  onClick={() => {
                    setRole(r);
                    if (r === 'ADMIN') {
                      setIsRegister(false);
                    }
                    window.history.pushState({}, '', `/auth/${r.toLowerCase()}`);
                  }}
                >
                  <div className="role-card-icon">
                    {getRoleIcon(r)}
                  </div>
                  <span className="role-title">{title}</span>
                </div>
              );
            })}
          </div>

          {role !== 'ADMIN' && (
            <div className="auth-mode-toggle">
              <button
                type="button"
                className={`auth-mode-btn ${!isRegister ? 'active' : ''}`}
                onClick={() => { setIsRegister(false); setErrorText(null); }}
              >
                Sign In
              </button>
              <button
                type="button"
                className={`auth-mode-btn ${isRegister ? 'active' : ''}`}
                onClick={() => { setIsRegister(true); setErrorText(null); }}
              >
                Register
              </button>
            </div>
          )}

          <form onSubmit={handleSubmit} style={{ display: 'flex', flexDirection: 'column', overflow: 'hidden', flex: 1 }}>
            
            <div className="form-scroll-area">
              {isRegister && (
                <>
                  <div className="form-group">
                    <label className="input-label">Full Name (पूरा नाम)</label>
                    <div className="input-wrapper">
                      <input 
                        type="text" 
                        className="form-input"
                        style={{ paddingLeft: '12px' }}
                        value={fullName}
                        onChange={(e) => setFullName(e.target.value)}
                        placeholder="Enter full name as per Aadhaar"
                        required
                      />
                    </div>
                  </div>

                  {role === 'CONTRACTOR' && (
                    <div className="form-group">
                      <label className="input-label">Company Name (फर्म का नाम)</label>
                      <div className="input-wrapper">
                        <input 
                          type="text" 
                          className="form-input"
                          style={{ paddingLeft: '12px' }}
                          value={companyName}
                          onChange={(e) => setCompanyName(e.target.value)}
                          placeholder="Enter contractor company name"
                          required
                        />
                      </div>
                    </div>
                  )}

                  {role === 'CUSTOMER' && (
                    <>
                      <div className="form-group">
                        <label className="input-label">Home Address (घर का पता)</label>
                        <div className="input-wrapper">
                          <input 
                            type="text" 
                            className="form-input"
                            style={{ paddingLeft: '12px' }}
                            value={homeAddress}
                            onChange={(e) => setHomeAddress(e.target.value)}
                            placeholder="Enter residential address"
                            required
                          />
                        </div>
                      </div>
                      <div className="form-group">
                        <label className="input-label">Preferred Language (पसंदीदा भाषा)</label>
                        <div className="input-wrapper">
                          <select 
                            className="form-input"
                            style={{ paddingLeft: '12px' }}
                            value={preferredLanguage}
                            onChange={(e) => setPreferredLanguage(e.target.value)}
                          >
                            <option value="en">English</option>
                            <option value="hi">Hindi (हिंदी)</option>
                            <option value="mr">Marathi (मराठी)</option>
                            <option value="ta">Tamil (தமிழ்)</option>
                            <option value="te">Telugu (తెలుగు)</option>
                            <option value="bn">Bengali (বাংলা)</option>
                          </select>
                        </div>
                      </div>
                    </>
                  )}

                  {role === 'WORKER' && (
                    <div className="form-group">
                      <label className="input-label">Select Your Skills (अपने हुनर चुनें)</label>
                      <div className="skills-container">
                        {Object.entries(groupedSkills).map(([cat, names]) => (
                          <div key={cat}>
                            <div className="skill-category">{cat}</div>
                            <div className="skill-grid">
                              {names.map(name => {
                                const isChecked = selectedSkills.includes(name);
                                return (
                                  <label key={name} className="checkbox-label" style={{ fontSize: '12px' }}>
                                    <input 
                                      type="checkbox"
                                      checked={isChecked}
                                      onChange={() => {
                                        if (isChecked) {
                                          setSelectedSkills(selectedSkills.filter(s => s !== name));
                                        } else {
                                          setSelectedSkills([...selectedSkills, name]);
                                        }
                                      }}
                                    />
                                    {name}
                                  </label>
                                );
                              })}
                            </div>
                          </div>
                        ))}
                      </div>
                    </div>
                  )}

                  <div className="form-group">
                    <label className="input-label">Email Address (वैकल्पिक)</label>
                    <div className="input-wrapper">
                      <input 
                        type="email" 
                        className="form-input"
                        style={{ paddingLeft: '12px' }}
                        value={email}
                        onChange={(e) => setEmail(e.target.value)}
                        placeholder="email@example.com"
                      />
                    </div>
                  </div>
                </>
              )}

              <div className="form-group">
                <label className="input-label">Phone Number (मोबाइल नंबर)</label>
                <div className="input-wrapper">
                  <Phone className="input-icon" size={16} />
                  <input 
                    type="tel" 
                    className="form-input"
                    value={phone}
                    onChange={(e) => setPhone(e.target.value)}
                    placeholder="Enter 10-digit mobile number"
                    required
                  />
                </div>
              </div>

              <div className="form-group">
                <label className="input-label">Password (पासवर्ड)</label>
                <div className="input-wrapper">
                  <Lock className="input-icon" size={16} />
                  <input 
                    type={showPassword ? "text" : "password"} 
                    className="form-input"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                    placeholder="Enter your password"
                    required
                  />
                  <button type="button" className="input-action" onClick={() => setShowPassword(!showPassword)}>
                    {showPassword ? <EyeOff size={16} /> : <Eye size={16} />}
                  </button>
                </div>
              </div>

              {!isRegister && (
                <div className="form-options">
                  <label className="checkbox-label">
                    <input 
                      type="checkbox" 
                      checked={rememberMe}
                      onChange={(e) => setRememberMe(e.target.checked)}
                    />
                    Remember me
                  </label>
                  <a href="#" className="forgot-link">Forgot Password?</a>
                </div>
              )}
              
              {errorText && (
                <div className="error-message">
                  {errorText}
                </div>
              )}
            </div>

            <div style={{ marginTop: '14px' }}>
              <button 
                type="submit" 
                className="btn-primary"
                disabled={loading}
              >
                <Lock size={16} />
                {loading ? 'Processing...' : (isRegister ? 'Register & Verify Identity' : 'Secure Login')}
              </button>
            </div>
            
          </form>

          <div className="auth-footer">
            <div className="security-badge">
              <Shield size={12} />
              Protected by LabourLink Trust Security
            </div>
            <div className="card-trust-strip">
              <div className="card-trust-item">
                <Shield size={12} className="card-trust-icon" />
                <span className="card-trust-text">Government Compliant</span>
              </div>
              <div className="card-trust-item">
                <Users size={12} className="card-trust-icon" />
                <span className="card-trust-text">Trusted by Thousands</span>
              </div>
              <div className="card-trust-item">
                <Star size={12} className="card-trust-icon" />
                <span className="card-trust-text">AI Powered Matching</span>
              </div>
            </div>
          </div>
        </div>

      </div>
    </div>
  );
};

export default Login;
