// web/src/pages/worker/Dashboard.tsx
import React, { useState, useEffect } from 'react';
import { User, WorkerProfile, VerificationRecord, Endorsement, Notification } from 'shared-types';
import TrustScoreGauge from '../../components/trust/TrustScoreGauge';
import VoiceControl from '../../components/shared/VoiceControl';
import EmptyState from '../../components/shared/EmptyState';
import { Clock, XCircle, ShieldCheck, MapPin, Radio, Bell, Briefcase } from 'lucide-react';
import ActivityTimeline from '../../components/ActivityTimeline';

interface WorkerDashboardProps {
  user: User;
  profileId: string;
  onLogout: () => void;
}

interface WorkerAssignment {
  id: string;
  booking_id: string;
  status: string;
  remarks: string | null;
  assigned_at: string;
  scheduled_start: string;
  booking_status: string;
  requirement_text: string;
  contractor_company: string | null;
  contractor_name: string | null;
}

export const WorkerDashboard: React.FC<WorkerDashboardProps> = ({ user, profileId, onLogout }) => {
  const [profile, setProfile] = useState<WorkerProfile | null>(null);
  const [trustDetails, setTrustDetails] = useState<any | null>(null);
  const [bookings, setBookings] = useState<any[]>([]);
  const [assignments, setAssignments] = useState<WorkerAssignment[]>([]);
  const [verifications, setVerifications] = useState<VerificationRecord[]>([]);
  const [endorsements, setEndorsements] = useState<Endorsement[]>([]);
  const [notifications, setNotifications] = useState<Notification[]>([]);
  const [opportunities, setOpportunities] = useState<any[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  // Engagement & Negotiation States
  const [engagements, setEngagements] = useState<any[]>([]);
  const [counterProposalAmount, setCounterProposalAmount] = useState<{ [engagementId: string]: string }>({});

  // Verification upload simulation
  const [uploading, setUploading] = useState<boolean>(false);
  const [docType, setDocType] = useState<'ID_DOCUMENT' | 'SKILL_CERT'>('ID_DOCUMENT');
  const [docUrl, setDocUrl] = useState<string>('');

  useEffect(() => {
    fetchDashboardData();
  }, [profileId]);

  const fetchDashboardData = async () => {
    setLoading(true);
    try {
      // 1. Fetch Profile
      const profRes = await fetch(`/api/workers/${profileId}/profile`);
      if (!profRes.ok) throw new Error('Failed to fetch worker profile');
      const profData = await profRes.json();
      setProfile(profData);

      // 2. Fetch Trust Score
      const trustRes = await fetch(`/api/workers/${profileId}/trust-score`);
      if (trustRes.ok) {
        const trustData = await trustRes.json();
        setTrustDetails(trustData.status === 'NOT_YET_ESTABLISHED' ? null : trustData);
      }

      // 3. Fetch Bookings (Replacing jobs with bookings)
      const bookingsRes = await fetch(`/api/bookings?userId=${profileId}&role=WORKER`);
      if (bookingsRes.ok) {
        const bookingsData = await bookingsRes.json();
        setBookings(bookingsData);
      }

      // 4. Fetch Assignments
      const assignmentsRes = await fetch(`/api/workers/${profileId}/assignments`);
      if (assignmentsRes.ok) {
        const assignmentsData = await assignmentsRes.json();
        setAssignments(assignmentsData);
      }

      // 5. Fetch Verifications
      const verifRes = await fetch(`/api/workers/${profileId}/verification-records`);
      if (verifRes.ok) {
        const verifData = await verifRes.json();
        setVerifications(verifData);
      }

      // 6. Fetch Notifications
      const notifRes = await fetch(`/api/notifications?userId=${user.id}`);
      if (notifRes.ok) {
        const notifData = await notifRes.json();
        setNotifications(notifData);
      }

      // 7. Fetch Endorsements
      const endorsementsRes = await fetch(`/api/workers/${profileId}/endorsements`).catch(() => null);
      if (endorsementsRes && endorsementsRes.ok) {
        const endData = await endorsementsRes.json();
        setEndorsements(Array.isArray(endData) ? endData : (endData.data || []));
      }

      // Fetch Opportunities
      const oppRes = await fetch(`/api/workers/${profileId}/opportunities`);
      if (oppRes.ok) {
        const oppData = await oppRes.json();
        setOpportunities(oppData);
      }

      // Fetch Engagements & Negotiations
      const engagementsRes = await fetch(`/api/engagements?userId=${user.id}&role=WORKER`);
      if (engagementsRes.ok) {
        const engagementsData = await engagementsRes.json();
        setEngagements(engagementsData);
      }

    } catch (err: any) {
      console.error(err);
    } finally {
      setLoading(false);
    }
  };

  const handleAvailabilityToggle = async (status: 'AVAILABLE' | 'BUSY' | 'UNAVAILABLE') => {
    try {
      const response = await fetch(`/api/workers/${profileId}/availability`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ status, set_via: 'UI' })
      });

      if (!response.ok) throw new Error('Failed to update availability');
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleProposePriceOffer = async (engagementId: string) => {
    const amountStr = counterProposalAmount[engagementId];
    if (!amountStr || isNaN(parseFloat(amountStr))) {
      alert('Please enter a valid counter-offer amount.');
      return;
    }

    try {
      const res = await fetch(`/api/engagements/${engagementId}/propose`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          offered_by: user.id,
          amount: parseFloat(amountStr),
          note: 'Worker Counter Proposal'
        })
      });

      if (!res.ok) throw new Error('Failed to submit counter offer');
      alert('Counter offer proposed successfully!');
      setCounterProposalAmount(prev => ({ ...prev, [engagementId]: '' }));
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleRespondToProposal = async (engagementId: string, response: 'ACCEPTED' | 'REJECTED') => {
    try {
      const res = await fetch(`/api/engagements/${engagementId}/respond`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ response })
      });

      if (!res.ok) throw new Error('Failed to respond to proposal');
      alert(`Proposal ${response.toLowerCase()} successfully!`);
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleDocSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!docUrl) return;

    setUploading(true);
    try {
      const response = await fetch(`/api/workers/${profileId}/verification-records`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ type: docType, evidence_url: docUrl })
      });

      if (!response.ok) throw new Error('Failed to submit document');
      
      setDocUrl('');
      alert('Document submitted and processed by Verification Agent successfully.');
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    } finally {
      setUploading(false);
    }
  };

  const handleRespondToAssignment = async (assignmentId: string, response: 'ACCEPTED' | 'REJECTED') => {
    try {
      const res = await fetch(`/api/assignments/${assignmentId}/respond`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ response })
      });
      if (!res.ok) throw new Error('Failed to respond to assignment');
      alert(`Assignment ${response.toLowerCase()} successfully!`);
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleAcceptBooking = async (bookingId: string) => {
    try {
      const response = await fetch(`/api/bookings/${bookingId}/accept`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' }
      });
      if (!response.ok) throw new Error('Failed to accept booking');
      
      alert('Booking accepted. Your status has been set to BUSY.');
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleAcceptOpportunity = async (opp: any) => {
    try {
      if (opp.type === 'CUSTOMER_BOOKING') {
        const bookRes = await fetch('/api/bookings', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            service_request_id: opp.id,
            worker_id: profileId,
            customer_id: opp.customer_id
          })
        });
        if (!bookRes.ok) throw new Error('Failed to create booking for opportunity.');
        const data = await bookRes.json();
        const acceptRes = await fetch(`/api/bookings/${data.bookingId}/accept`, {
          method: 'PATCH',
          headers: { 'Content-Type': 'application/json' }
        });
        if (!acceptRes.ok) throw new Error('Failed to accept booking.');
        alert('Ghar ka Kaam (Booking) accepted successfully! Your status is now BUSY.');
      } else {
        const jobRes = await fetch('/api/jobs/offers', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            job_requirement_id: opp.id,
            worker_id: profileId,
            contractor_id: opp.contractor_id
          })
        });
        if (!jobRes.ok) throw new Error('Failed to apply for contractor job.');
        const data = await jobRes.json();
        const acceptRes = await fetch(`/api/jobs/${data.jobId}/accept`, {
          method: 'PATCH',
          headers: { 'Content-Type': 'application/json' }
        });
        if (!acceptRes.ok) throw new Error('Failed to accept job offer.');
        alert('Contractor Job accepted successfully! Your status is now BUSY.');
      }
      fetchDashboardData();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleReadNotification = async (notifId: string) => {
    try {
      await fetch(`/api/notifications/${notifId}/read`, { method: 'PATCH' });
      fetchDashboardData();
    } catch (e) {}
  };

  if (loading) {
    return <div className="container" style={{ textAlign: 'center', padding: '60px', fontFamily: 'Outfit, sans-serif' }}>Loading Official Profile Data...</div>;
  }

  return (
    <div style={{ fontFamily: 'Outfit, sans-serif', backgroundColor: '#f1f5f9', minHeight: '100vh', padding: '24px' }}>
      {/* Header Bar */}
      <header style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', backgroundColor: '#ffffff', padding: '16px 24px', border: '1px solid #e2e8f0', borderRadius: '12px', marginBottom: '24px', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
        <div style={{ display: 'flex', alignItems: 'center', gap: '12px' }}>
          <ShieldCheck size={28} color="#0f766e" />
          <div>
            <h1 style={{ margin: 0, fontSize: '20px', color: '#0f766e', fontWeight: 'bold' }}>Workforce Hub (श्रमिक पोर्टल)</h1>
            <p style={{ margin: '2px 0 0 0', fontSize: '13px', color: '#64748b' }}>
              Logged in: <strong style={{ color: '#1e293b' }}>{user.full_name}</strong> | UID: {profile?.id.slice(0, 8)}
            </p>
          </div>
        </div>
        <button 
          onClick={onLogout} 
          style={{ backgroundColor: '#ffffff', color: '#ef4444', border: '1px solid #fecaca', padding: '8px 16px', fontSize: '13px', fontWeight: '600', borderRadius: '8px', cursor: 'pointer' }}
        >
          Logout Session
        </button>
      </header>

      <div className="container" style={{ maxWidth: '1400px', margin: '0 auto' }}>
        {/* Verification Status Banner */}
        <div className="alert-banner alert-banner-info" style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', backgroundColor: '#e0f2fe', border: '1px solid #bae6fd', padding: '14px 20px', borderRadius: '10px', marginBottom: '24px', color: '#0369a1' }}>
          <div style={{ display: 'flex', gap: '10px', alignItems: 'center' }}>
            <MapPin size={20} />
            <div>
              <strong>Verification Badge:</strong> Your profile is currently marked as: <strong style={{ color: '#0369a1' }}>{profile?.verification_status}</strong>
              <div style={{ fontSize: '12px', color: '#0284c7', marginTop: '2px' }}>
                Home Location: {profile?.home_lat.toFixed(4)} N, {profile?.home_lng.toFixed(4)} E
              </div>
            </div>
          </div>
          <div style={{ fontWeight: 'bold', textTransform: 'uppercase', fontSize: '11px', padding: '4px 8px', borderRadius: '6px', backgroundColor: profile?.verification_status === 'VERIFIED' ? '#d1fae5' : '#fef3c7', color: profile?.verification_status === 'VERIFIED' ? '#065f46' : '#92400e' }}>
            {profile?.verification_status === 'VERIFIED' ? 'Approved' : 'Pending Review'}
          </div>
        </div>

        <div style={{ display: 'grid', gridTemplateColumns: '1fr 2.5fr', gap: '24px' }}>
          
          {/* SIDEBAR */}
          <div>
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '20px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '14px', fontWeight: 'bold', color: '#1e293b' }}>Availability Status</h4>
              <div style={{ display: 'flex', flexDirection: 'column', gap: '10px' }}>
                <button 
                  className={`btn ${profile?.availability_status === 'AVAILABLE' ? 'btn-primary' : 'btn-secondary'}`}
                  onClick={() => handleAvailabilityToggle('AVAILABLE')}
                  style={{ display: 'flex', alignItems: 'center', gap: '8px', padding: '10px', borderRadius: '8px', border: 'none', cursor: 'pointer', fontWeight: '600', backgroundColor: profile?.availability_status === 'AVAILABLE' ? '#0f766e' : '#f1f5f9', color: profile?.availability_status === 'AVAILABLE' ? '#ffffff' : '#475569' }}
                >
                  <Radio size={16} /> Set Available (काम के लिए उपलब्ध)
                </button>
                <button 
                  className={`btn ${profile?.availability_status === 'BUSY' ? 'btn-accent' : 'btn-secondary'}`}
                  onClick={() => handleAvailabilityToggle('BUSY')}
                  style={{ display: 'flex', alignItems: 'center', gap: '8px', padding: '10px', borderRadius: '8px', border: 'none', cursor: 'pointer', fontWeight: '600', backgroundColor: profile?.availability_status === 'BUSY' ? '#d97706' : '#f1f5f9', color: profile?.availability_status === 'BUSY' ? '#ffffff' : '#475569' }}
                >
                  <Clock size={16} /> Set Busy (काम पर व्यस्त)
                </button>
                <button 
                  className={`btn ${profile?.availability_status === 'UNAVAILABLE' ? 'btn-danger' : 'btn-secondary'}`}
                  onClick={() => handleAvailabilityToggle('UNAVAILABLE')}
                  style={{ display: 'flex', alignItems: 'center', gap: '8px', padding: '10px', borderRadius: '8px', border: 'none', cursor: 'pointer', fontWeight: '600', backgroundColor: profile?.availability_status === 'UNAVAILABLE' ? '#ef4444' : '#f1f5f9', color: profile?.availability_status === 'UNAVAILABLE' ? '#ffffff' : '#475569' }}
                >
                  <XCircle size={16} /> Set Offline (उपलब्ध नहीं)
                </button>
              </div>
            </div>

            <TrustScoreGauge 
              score={trustDetails?.score ?? null} 
              version={trustDetails?.version}
              contributingFactors={trustDetails?.contributing_factors}
            />

            {/* Document Submission */}
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '20px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginTop: '24px' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '14px', fontWeight: 'bold' }}>Audit Verification File</h4>
              <form onSubmit={handleDocSubmit}>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label" style={{ fontSize: '12px', fontWeight: '600', display: 'block', marginBottom: '6px' }}>File Type</label>
                  <select className="form-control" value={docType} onChange={(e) => setDocType(e.target.value as any)} style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}>
                    <option value="ID_DOCUMENT">Aadhaar / ID Card</option>
                    <option value="SKILL_CERT">Skill Certificate</option>
                  </select>
                </div>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label" style={{ fontSize: '12px', fontWeight: '600', display: 'block', marginBottom: '6px' }}>Document URL Link</label>
                  <input 
                    type="text" 
                    className="form-control" 
                    placeholder="https://docs.link/file.pdf"
                    value={docUrl}
                    onChange={(e) => setDocUrl(e.target.value)}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                    required 
                  />
                </div>
                <button type="submit" className="btn btn-primary" style={{ width: '100%', padding: '10px', borderRadius: '8px', border: 'none', backgroundColor: '#0f766e', color: '#ffffff', fontWeight: 'bold', cursor: 'pointer' }} disabled={uploading}>
                  {uploading ? 'Scanning...' : 'Submit to Verification Agent'}
                </button>
              </form>

              {verifications.length > 0 && (
                <div style={{ marginTop: '16px', borderTop: '1px dashed #cbd5e1', paddingTop: '12px' }}>
                  <h5 style={{ fontSize: '12px', fontWeight: 'bold', color: '#475569', marginBottom: '8px' }}>Submitted Files Checklist</h5>
                  <div style={{ display: 'flex', flexDirection: 'column', gap: '6px' }}>
                    {verifications.map(v => (
                      <div key={v.id} style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', fontSize: '11px', padding: '6px 10px', border: '1px solid #e2e8f0', backgroundColor: '#f8fafc', borderRadius: '6px' }}>
                        <span style={{ textTransform: 'capitalize', fontWeight: '500' }}>{v.type.toLowerCase().replace('_', ' ')}</span>
                        <span style={{ fontSize: '9px', fontWeight: 'bold', padding: '2px 6px', borderRadius: '4px', backgroundColor: v.status === 'VERIFIED' ? '#d1fae5' : '#fef3c7', color: v.status === 'VERIFIED' ? '#065f46' : '#92400e' }}>
                          {v.status}
                        </span>
                      </div>
                    ))}
                  </div>
                </div>
              )}
            </div>

            <div className="card" style={{ backgroundColor: '#ffffff', padding: '20px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginTop: '24px' }}>
              <ActivityTimeline userId={user.id} />
            </div>
          </div>

          {/* MAIN CONTENT */}
          <div>
            {/* Voice Control panel */}
            <VoiceControl userId={user.id} onCommandProcessed={() => fetchDashboardData()} />

            {/* Contractor Project Assignments (New Section) */}
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px', borderLeft: '4px solid #0f766e' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '16px', fontWeight: '700', color: '#0f766e', display: 'flex', alignItems: 'center', gap: '8px' }}>
                <Briefcase size={20} /> Contractor Project Assignments (नियोजक काम ऑफर)
              </h4>
              {assignments.length === 0 ? (
                <div style={{ padding: '24px', textAlign: 'center', fontSize: '13px', color: '#64748b', border: '1px dashed #cbd5e1', borderRadius: '8px' }}>
                  No contractor project assignments offered.
                </div>
              ) : (
                <div style={{ display: 'flex', flexDirection: 'column', gap: '12px' }}>
                  {assignments.map((asn) => (
                    <div key={asn.id} style={{ padding: '16px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#f8fafc' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start' }}>
                        <div>
                          <span style={{ fontSize: '10px', color: '#94a3b8', fontWeight: '500' }}>Assignment ID: {asn.id.slice(0, 8)}...</span>
                          <h5 style={{ margin: '4px 0', fontSize: '14px', fontWeight: '700', color: '#1e293b' }}>
                            {asn.contractor_company ? `${asn.contractor_company} (${asn.contractor_name})` : 'Independent Contractor'}
                          </h5>
                          <p style={{ margin: '6px 0', fontSize: '12px', color: '#475569' }}>
                            Project Details: <strong>"{asn.requirement_text}"</strong>
                          </p>
                          {asn.remarks && (
                            <p style={{ margin: '2px 0', fontSize: '11px', color: '#64748b', fontStyle: 'italic' }}>
                              Remarks: {asn.remarks}
                            </p>
                          )}
                        </div>
                        <div style={{ textAlign: 'right' }}>
                          <span style={{ 
                            fontSize: '11px', 
                            fontWeight: 'bold', 
                            padding: '3px 8px', 
                            borderRadius: '6px',
                            backgroundColor: asn.status === 'ACCEPTED' ? '#d1fae5' : asn.status === 'ASSIGNED' ? '#fef3c7' : '#fee2e2',
                            color: asn.status === 'ACCEPTED' ? '#065f46' : asn.status === 'ASSIGNED' ? '#92400e' : '#991b1b'
                          }}>
                            {asn.status}
                          </span>
                        </div>
                      </div>

                      {asn.status === 'ASSIGNED' && (
                        <div style={{ display: 'flex', gap: '8px', justifyContent: 'flex-end', marginTop: '12px', borderTop: '1px solid #e2e8f0', paddingTop: '10px' }}>
                          <button 
                            onClick={() => handleRespondToAssignment(asn.id, 'REJECTED')}
                            style={{ backgroundColor: '#ffffff', color: '#ef4444', border: '1px solid #fecaca', padding: '6px 12px', fontSize: '12px', fontWeight: '600', borderRadius: '6px', cursor: 'pointer' }}
                          >
                            Decline
                          </button>
                          <button 
                            onClick={() => handleRespondToAssignment(asn.id, 'ACCEPTED')}
                            style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: '600', borderRadius: '6px', cursor: 'pointer' }}
                          >
                            Accept & Go Busy
                          </button>
                        </div>
                      )}
                    </div>
                  ))}
                </div>
              )}
            </div>

            {/* Opportunities Feed */}
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px', borderLeft: '4px solid #2563eb' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '16px', fontWeight: '700', color: '#2563eb' }}>Opportunity Feed (नये काम के सुझाव)</h4>
              {opportunities.length === 0 ? (
                <div style={{ padding: '20px', textAlign: 'center', fontSize: '13px', color: '#64748b', border: '1px dashed #cbd5e1', borderRadius: '8px' }}>
                  No new opportunities detected nearby. Try setting your availability status to AVAILABLE.
                </div>
              ) : (
                <div style={{ display: 'flex', flexDirection: 'column', gap: '12px' }}>
                  {opportunities.map((opp) => (
                    <div key={opp.id} style={{ padding: '12px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#f8fafc' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start' }}>
                        <div>
                          <span style={{ fontSize: '10px', fontWeight: 'bold', padding: '2px 6px', borderRadius: '4px', backgroundColor: opp.type === 'CUSTOMER_BOOKING' ? '#d1fae5' : '#dbeafe', color: opp.type === 'CUSTOMER_BOOKING' ? '#065f46' : '#1e40af', marginBottom: '4px', display: 'inline-block' }}>
                            {opp.type === 'CUSTOMER_BOOKING' ? 'Residential Booking' : 'Contractor Project'}
                          </span>
                          <h5 style={{ margin: '4px 0', fontSize: '14px', fontWeight: 'bold' }}>{opp.title}</h5>
                          <p style={{ margin: '4px 0', fontSize: '12px', color: '#475569' }}>{opp.description}</p>
                          <p style={{ margin: '4px 0', fontSize: '11px', color: '#0f766e', fontStyle: 'italic' }}><strong>Reasoning:</strong> {opp.reasoning}</p>
                        </div>
                        <div style={{ textAlign: 'right', minWidth: '90px' }}>
                          <span style={{ fontSize: '13px', fontWeight: 'bold', color: '#0f766e' }}>{opp.match_score}% Match</span>
                          <div style={{ fontSize: '11px', color: '#64748b', marginTop: '2px' }}>{opp.distance_km} km away</div>
                          <div style={{ fontSize: '12px', fontWeight: 'bold', color: '#d97706', marginTop: '4px' }}>Est. Rs. {opp.estimated_earnings}</div>
                        </div>
                      </div>
                      <div style={{ display: 'flex', justifyContent: 'flex-end', marginTop: '8px' }}>
                        <button 
                          className="btn btn-primary" 
                          style={{ padding: '6px 12px', fontSize: '11px', backgroundColor: '#0f766e', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer', fontWeight: 'bold' }}
                          onClick={() => handleAcceptOpportunity(opp)}
                        >
                          Accept Opportunity
                        </button>
                      </div>
                    </div>
                  ))}
                </div>
              )}
            </div>

            {/* In-App Notifications Feed */}
            {notifications.length > 0 && (
              <div className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px', borderLeft: '4px solid #ea580c' }}>
                <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '16px', fontWeight: '700', color: '#ea580c', display: 'flex', alignItems: 'center', gap: '6px' }}>
                  <Bell size={18} /> Alert & Dispatch Feed
                </h4>
                <div style={{ display: 'flex', flexDirection: 'column', gap: '10px' }}>
                  {notifications.map(notif => (
                    <div 
                      key={notif.id} 
                      style={{ 
                        padding: '12px', 
                        border: '1px solid #e2e8f0', 
                        backgroundColor: notif.read_at ? '#ffffff' : '#fffbeb',
                        borderRadius: '8px',
                        display: 'flex',
                        justifyContent: 'space-between',
                        alignItems: 'center'
                      }}
                    >
                      <div>
                        <span style={{ fontSize: '9px', fontWeight: 'bold', padding: '2px 6px', borderRadius: '4px', backgroundColor: '#f1f5f9', color: '#475569', marginRight: '6px' }}>{notif.type}</span>
                        <span style={{ fontSize: '13px', fontWeight: notif.read_at ? 'normal' : 'bold', color: '#1e293b' }}>{notif.payload?.title || 'Notification'}</span>
                        <p style={{ margin: '4px 0 0 0', fontSize: '12px', color: '#475569' }}>{notif.payload?.message || ''}</p>
                        <div style={{ fontSize: '11px', color: '#94a3b8', marginTop: '4px' }}>Channel: {notif.channel}</div>
                      </div>
                      {!notif.read_at && (
                        <button 
                          className="btn btn-secondary" 
                          style={{ padding: '4px 8px', fontSize: '11px', borderRadius: '6px', cursor: 'pointer' }}
                          onClick={() => handleReadNotification(notif.id)}
                        >
                          Mark Read
                        </button>
                      )}
                    </div>
                  ))}
                </div>
              </div>
            )}

            {/* Active Contracts & Negotiations */}
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '16px', fontWeight: '700', color: '#1e293b', display: 'flex', alignItems: 'center', gap: '8px' }}>
                🤝 ACTIVE CONTRACTS & NEGOTIATIONS (सक्रिय सौदेबाजी और बातचीत)
              </h4>
              {engagements.length === 0 ? (
                <div style={{ padding: '24px', textAlign: 'center', fontSize: '13px', color: '#64748b', border: '1px dashed #cbd5e1', borderRadius: '8px' }}>
                  No active negotiations or engagements found.
                </div>
              ) : (
                <div style={{ display: 'flex', flexDirection: 'column', gap: '12px' }}>
                  {engagements.map((eng) => {
                    const lastOffer = eng.offers && eng.offers[0];
                    const isPendingOrNegotiating = eng.status === 'PENDING' || eng.status === 'NEGOTIATING';
                    const lastOfferedByMe = lastOffer && lastOffer.offered_by === user.id;

                    return (
                      <div key={eng.id} style={{ padding: '16px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#f8fafc' }}>
                        <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: '8px' }}>
                          <div>
                            <span style={{ fontSize: '10px', color: '#94a3b8', fontWeight: '500' }}>ENGAGEMENT ID: {eng.id.slice(0, 8)}...</span>
                            <h5 style={{ margin: '4px 0', fontSize: '14px', fontWeight: '700', color: '#1e293b' }}>
                              Client/Initiator: {eng.initiator_name} ({eng.mode === 'DIRECT_WORKER' ? '👤 Direct Customer' : '🏢 Contractor Staffing'})
                            </h5>
                            <p style={{ margin: '6px 0', fontSize: '12px', color: '#475569' }}>
                              Requirement Description: <strong>"{eng.request_text}"</strong>
                            </p>
                          </div>
                          <span style={{
                            fontSize: '11px',
                            fontWeight: 'bold',
                            padding: '4px 8px',
                            borderRadius: '6px',
                            backgroundColor: eng.status === 'COMPLETED' ? '#e2f5f3' : (eng.status === 'IN_PROGRESS' ? '#dbeafe' : (eng.status === 'DISPUTED' ? '#fee2e2' : '#f1f5f9')),
                            color: eng.status === 'COMPLETED' ? '#0f766e' : (eng.status === 'IN_PROGRESS' ? '#2563eb' : (eng.status === 'DISPUTED' ? '#dc2626' : '#475569'))
                          }}>
                            {eng.status}
                          </span>
                        </div>

                        {/* Offers Log */}
                        {eng.offers && eng.offers.length > 0 && (
                          <div style={{ marginTop: '10px', padding: '10px', backgroundColor: '#ffffff', borderRadius: '6px', border: '1px solid #e2e8f0' }}>
                            <span style={{ fontSize: '11px', fontWeight: 'bold', color: '#64748b', display: 'block', marginBottom: '4px' }}>Negotiation Log:</span>
                            <div style={{ display: 'flex', flexDirection: 'column', gap: '4px' }}>
                              {eng.offers.slice(0, 3).map((o: any) => (
                                <div key={o.id} style={{ display: 'flex', justifyContent: 'space-between', fontSize: '12px', color: '#475569' }}>
                                  <span>₹{o.amount} proposed by {o.offered_by_name}</span>
                                  <span style={{ fontSize: '10px', color: '#94a3b8' }}>{new Date(o.created_at).toLocaleTimeString()}</span>
                                </div>
                              ))}
                            </div>
                          </div>
                        )}

                        {/* Controls */}
                        <div style={{ marginTop: '12px', display: 'flex', flexWrap: 'wrap', gap: '8px', alignItems: 'center' }}>
                          {isPendingOrNegotiating && (
                            <>
                              {!lastOfferedByMe ? (
                                <>
                                  <button
                                    onClick={() => handleRespondToProposal(eng.id, 'ACCEPTED')}
                                    style={{ backgroundColor: '#10b981', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                  >
                                    Accept Offer (₹{lastOffer?.amount || 500})
                                  </button>
                                  <button
                                    onClick={() => handleRespondToProposal(eng.id, 'REJECTED')}
                                    style={{ backgroundColor: '#ef4444', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                  >
                                    Reject Offer
                                  </button>
                                </>
                              ) : (
                                <span style={{ fontSize: '12px', color: '#64748b', fontStyle: 'italic' }}>Awaiting counterparty response...</span>
                              )}

                              <div style={{ display: 'flex', gap: '4px', marginLeft: 'auto' }}>
                                <input
                                  type="number"
                                  placeholder="Counter offer price"
                                  value={counterProposalAmount[eng.id] || ''}
                                  onChange={(e) => setCounterProposalAmount({ ...counterProposalAmount, [eng.id]: e.target.value })}
                                  style={{ width: '130px', padding: '5px 8px', fontSize: '12px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                                />
                                <button
                                  onClick={() => handleProposePriceOffer(eng.id)}
                                  style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                >
                                  Counter
                                </button>
                              </div>
                            </>
                          )}

                          {eng.status === 'ACCEPTED' && (
                            <span style={{ fontSize: '12px', color: '#2563eb', fontStyle: 'italic', fontWeight: '600' }}>
                              ✓ Offer Accepted. Awaiting client to trigger start of work.
                            </span>
                          )}

                          {eng.status === 'IN_PROGRESS' && (
                            <span style={{ fontSize: '12px', color: '#d97706', fontStyle: 'italic', fontWeight: '600' }}>
                              ⚡ Job in Progress. Set your availability status to BUSY.
                            </span>
                          )}

                          {eng.status === 'DISPUTED' && (
                            <div style={{ color: '#dc2626', fontSize: '12px', fontWeight: '600' }}>
                              ⚠️ Dispute under review. Your status is temporarily locked to busy.
                            </div>
                          )}
                        </div>
                      </div>
                    );
                  })}
                </div>
              )}
            </div>

            {/* Jobs Timeline */}
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '16px', fontWeight: '700', color: '#1e293b' }}>Secured Job History (रोजगार विवरण)</h4>
              {bookings.length === 0 ? (
                <EmptyState variant="default" message="No jobs found in your history queue. Once a customer or contractor assigns you a job, it will appear here." />
              ) : (
                <table className="table" style={{ width: '100%', borderCollapse: 'collapse', marginTop: '12px' }}>
                  <thead>
                    <tr style={{ textAlign: 'left', borderBottom: '1px solid #cbd5e1' }}>
                      <th style={{ padding: '10px', fontSize: '12px', color: '#64748b' }}>Project / Client</th>
                      <th style={{ padding: '10px', fontSize: '12px', color: '#64748b' }}>Scheduled Date</th>
                      <th style={{ padding: '10px', fontSize: '12px', color: '#64748b' }}>Amount</th>
                      <th style={{ padding: '10px', fontSize: '12px', color: '#64748b' }}>Payment</th>
                      <th style={{ padding: '10px', fontSize: '12px', color: '#64748b' }}>Job Status</th>
                      <th style={{ padding: '10px', fontSize: '12px', color: '#64748b' }}>Actions</th>
                    </tr>
                  </thead>
                  <tbody>
                    {bookings.map((b: any) => (
                      <tr key={b.id} style={{ borderBottom: '1px solid #f1f5f9' }}>
                        <td style={{ padding: '12px 10px' }}>
                          <strong style={{ fontSize: '13px', color: '#1e293b' }}>{b.contractor_company ? `🏢 ${b.contractor_company}` : `👤 Customer: ${b.customer_name}`}</strong>
                          <div style={{ fontSize: '11px', color: '#64748b', marginTop: '2px' }}>Requirement: {b.requirement_text ? b.requirement_text.slice(0, 50) : ''}...</div>
                        </td>
                        <td style={{ padding: '12px 10px', fontSize: '12px', color: '#475569' }}>{new Date(b.scheduled_start).toLocaleDateString()}</td>
                        <td style={{ padding: '12px 10px', fontSize: '12px', color: '#475569', fontWeight: 'bold' }}>Rs. {b.payment_amount || 'N/A'}</td>
                        <td style={{ padding: '12px 10px' }}>
                          <span style={{ fontSize: '10px', fontWeight: 'bold', padding: '2px 6px', borderRadius: '4px', backgroundColor: b.payment_status === 'CONFIRMED' ? '#d1fae5' : '#fef3c7', color: b.payment_status === 'CONFIRMED' ? '#065f46' : '#92400e' }}>
                            {b.payment_status || 'PENDING'}
                          </span>
                        </td>
                        <td style={{ padding: '12px 10px' }}>
                          <span style={{ fontSize: '10px', fontWeight: 'bold', padding: '2px 6px', borderRadius: '4px', backgroundColor: b.status === 'COMPLETED' ? '#d1fae5' : (b.status === 'ACCEPTED' || b.status === 'IN_PROGRESS' ? '#fef3c7' : '#f1f5f9'), color: b.status === 'COMPLETED' ? '#065f46' : (b.status === 'ACCEPTED' || b.status === 'IN_PROGRESS' ? '#92400e' : '#475569') }}>
                            {b.status}
                          </span>
                        </td>
                        <td style={{ padding: '12px 10px' }}>
                          {b.status === 'REQUESTED' && (
                            <button 
                              onClick={() => handleAcceptBooking(b.id)}
                              style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '4px 8px', fontSize: '11px', borderRadius: '4px', cursor: 'pointer', fontWeight: 'bold' }}
                            >
                              Accept Offer
                            </button>
                          )}
                          {b.status === 'ACCEPTED' && (
                            <span style={{ fontSize: '11px', color: '#64748b' }}>Awaiting work start</span>
                          )}
                          {b.status === 'IN_PROGRESS' && (
                            <span style={{ fontSize: '11px', color: '#059669', fontWeight: 'bold' }}>Active Job Site</span>
                          )}
                          {b.status === 'COMPLETED' && (
                            <span style={{ fontSize: '11px', color: '#64748b' }}>Record closed</span>
                          )}
                        </td>
                      </tr>
                    ))}
                  </tbody>
                </table>
              )}
            </div>

            {/* Endorsements Audit */}
            <div className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
              <h4 className="card-title" style={{ margin: '0 0 16px 0', fontSize: '16px', fontWeight: '700', color: '#1e293b' }}>Verifiable Contractor Endorsements</h4>
              {endorsements.length === 0 ? (
                <div style={{ padding: '12px', textAlign: 'center', fontSize: '12px', color: '#64748b' }}>
                  No verified contractor endorsements recorded yet on your trust profile.
                </div>
              ) : (
                <div style={{ display: 'flex', flexDirection: 'column', gap: '10px' }}>
                  {endorsements.map((end: any) => (
                    <div key={end.id} style={{ padding: '12px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#f8fafc' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', fontWeight: 'bold', fontSize: '12px', color: '#0f766e', marginBottom: '4px' }}>
                        <span>Skill: {end.skill}</span>
                        <span style={{ color: '#64748b' }}>UID: {end.endorser_id.slice(0, 8)}</span>
                      </div>
                      <p style={{ fontSize: '13px', fontStyle: 'italic', margin: '4px 0' }}>"{end.comment}"</p>
                      <div style={{ fontSize: '10px', color: '#94a3b8', marginTop: '6px', textAlign: 'right' }}>
                        Recorded on: {new Date(end.created_at).toLocaleDateString()}
                      </div>
                    </div>
                  ))}
                </div>
              )}
            </div>
          </div>

        </div>
      </div>
    </div>
  );
};

export default WorkerDashboard;
