// web/src/pages/customer/Dashboard.tsx
import React, { useState, useEffect } from 'react';
import ActivityTimeline from '../../components/ActivityTimeline';
import TrustScoreBreakdownModal from '../../components/TrustScoreBreakdownModal';
import { User } from 'shared-types';

interface CustomerDashboardProps {
  user: User;
  profileId: string;
  onLogout: () => void;
}

interface Recommendation {
  worker_id: string;
  full_name: string;
  rank: number;
  explanation: string;
  match_score: number;
  trust_score: number | null;
  skills: string[];
  distance_km: number;
}

interface Booking {
  id: string;
  worker_id: string | null;
  worker_name: string;
  contractor_id: string | null;
  contractor_name: string;
  contractor_company: string;
  status: string;
  scheduled_start: string;
  requirement_text: string;
  payment_amount: number | null;
  payment_status: string | null;
}

interface Contractor {
  id: string;
  user_id: string;
  company_name: string;
  full_name: string;
  phone: string;
  email: string;
}

export const CustomerDashboard: React.FC<CustomerDashboardProps> = ({ user, profileId, onLogout }) => {
  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingAssignments, setBookingAssignments] = useState<{ [bookingId: string]: any[] }>({});
  const [rawText, setRawText] = useState<string>('');
  const [submitting, setSubmitting] = useState<boolean>(false);
  const [activeReqId, setActiveReqId] = useState<string | null>(null);
  const [recommendations, setRecommendations] = useState<Recommendation[]>([]);
  
  // Hiring Flow Mode: 'WORKER' or 'CONTRACTOR'
  const [hireMode, setHireMode] = useState<'WORKER' | 'CONTRACTOR'>('WORKER');
  const [contractors, setContractors] = useState<Contractor[]>([]);
  const [selectedContractorId, setSelectedContractorId] = useState<string>('');

  // New Engagement & Negotiation States
  const [engagements, setEngagements] = useState<any[]>([]);
  const [disputingEngagementId, setDisputingEngagementId] = useState<string | null>(null);
  const [disputeReason, setDisputeReason] = useState<string>('');
  const [disputePhotos, setDisputePhotos] = useState<string>('');
  const [counterProposalAmount, setCounterProposalAmount] = useState<{ [engagementId: string]: string }>({});

  // Modals / Overlays
  const [activeExplainWorkerId, setActiveExplainWorkerId] = useState<string | null>(null);
  const [ratingBookingId, setRatingBookingId] = useState<string | null>(null);
  const [ratingWorkerId, setRatingWorkerId] = useState<string | null>(null);
  const [ratingScore, setRatingScore] = useState<number>(5);
  const [ratingComment, setRatingComment] = useState<string>('');
  
  const [paymentBookingId, setPaymentBookingId] = useState<string | null>(null);
  const [paymentAmount, setPaymentAmount] = useState<string>('500');
  const [paymentMethod, setPaymentMethod] = useState<'UPI_VERIFIED' | 'CASH_ATTESTED' | 'BANK_VERIFIED'>('UPI_VERIFIED');

  useEffect(() => {
    fetchBookings();
    fetchContractors();
    fetchEngagements();
  }, []);

  const fetchEngagements = async () => {
    try {
      const res = await fetch(`/api/engagements?userId=${user.id}&role=CUSTOMER`);
      if (res.ok) {
        const data = await res.json();
        setEngagements(data);
      }
    } catch (e) {
      console.error(e);
    }
  };

  const fetchBookings = async () => {
    try {
      const res = await fetch(`/api/bookings?userId=${profileId}&role=CUSTOMER`);
      if (res.ok) {
        const data = await res.json();
        setBookings(data);
        // Fetch assignments for contractor-based bookings
        data.forEach((b: any) => {
          if (b.contractor_id) {
            fetchBookingAssignments(b.id);
          }
        });
      }
    } catch (e) {
      console.error(e);
    }
  };

  const fetchBookingAssignments = async (bookingId: string) => {
    try {
      const res = await fetch(`/api/bookings/${bookingId}/assignments`);
      if (res.ok) {
        const data = await res.json();
        setBookingAssignments(prev => ({ ...prev, [bookingId]: data }));
      }
    } catch (e) {
      console.error(e);
    }
  };

  const fetchContractors = async () => {
    try {
      const res = await fetch('/api/customer/find-contractors');
      if (res.ok) {
        const data = await res.json();
        setContractors(data);
        if (data.length > 0) {
          setSelectedContractorId(data[0].id);
        }
      }
    } catch (e) {
      console.error(e);
    }
  };

  const handleQuickAction = (actionType: 'plumber' | 'electrician' | 'painter' | 'carpenter') => {
    switch (actionType) {
      case 'plumber':
        setRawText('I need a plumber to fix a leaking pipe in my bathroom as soon as possible.');
        break;
      case 'electrician':
        setRawText('I need a skilled electrician for wiring installation and switchboard repairs.');
        break;
      case 'painter':
        setRawText('I need a painter to repaint my living room wall.');
        break;
      case 'carpenter':
        setRawText('I need a carpenter to fix a broken wooden door hinge.');
        break;
    }
  };

  const handleSubmitRequest = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!rawText.trim()) return;

    try {
      setSubmitting(true);
      const res = await fetch('/api/customer/service-requests', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ customer_id: profileId, raw_text: rawText })
      });

      if (!res.ok) throw new Error('Failed to submit service request');
      const data = await res.json();
      
      if (hireMode === 'WORKER') {
        setActiveReqId(data.serviceRequestId);
        setRecommendations(data.recommendations);
        alert('Request analyzed successfully! Recommended workers found.');
      } else {
        // Hire Contractor Mode: Directly book the selected contractor with this request
        if (!selectedContractorId) {
          alert('Please select a contractor to book first.');
          return;
        }
        await handleBookContractor(data.serviceRequestId, selectedContractorId);
      }
      setRawText('');
    } catch (err: any) {
      alert(err.message);
    } finally {
      setSubmitting(false);
    }
  };

  const handleBookWorker = async (workerId: string) => {
    if (!activeReqId) return;

    try {
      const res = await fetch(`/api/requests/${activeReqId}/engage`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          mode: 'DIRECT_WORKER',
          initiator_id: user.id,
          counterparty_id: workerId,
          initial_amount: 500,
          note: 'Initial hiring request for direct worker.'
        })
      });

      if (!res.ok) throw new Error('Engagement could not be created');
      alert('Worker hiring engagement initiated! Awaiting worker response.');
      setActiveReqId(null);
      setRecommendations([]);
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleBookContractor = async (serviceRequestId: string, contractorId: string) => {
    try {
      const res = await fetch(`/api/requests/${serviceRequestId}/engage`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          mode: 'VIA_CONTRACTOR',
          initiator_id: user.id,
          counterparty_id: contractorId,
          initial_amount: 1500,
          note: 'Initial contractor project engagement.'
        })
      });

      if (!res.ok) throw new Error('Contractor engagement failed');
      alert('Contractor project engagement initiated!');
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleProposePriceOffer = async (engagementId: string) => {
    const amountStr = counterProposalAmount[engagementId];
    if (!amountStr || isNaN(parseFloat(amountStr))) {
      alert('Please enter a valid counter-offer amount.');
      return;
    }

    try {
      const res = await fetch(`/api/engagements/${engagementId}/propose`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          offered_by: user.id,
          amount: parseFloat(amountStr),
          note: 'Customer Counter Proposal'
        })
      });

      if (!res.ok) throw new Error('Failed to submit counter offer');
      alert('Counter offer proposed successfully!');
      setCounterProposalAmount(prev => ({ ...prev, [engagementId]: '' }));
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleRespondToProposal = async (engagementId: string, response: 'ACCEPTED' | 'REJECTED') => {
    try {
      const res = await fetch(`/api/engagements/${engagementId}/respond`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ response })
      });

      if (!res.ok) throw new Error('Failed to respond to proposal');
      alert(`Proposal ${response.toLowerCase()} successfully!`);
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleUpdateEngagementStatus = async (engagementId: string, status: string) => {
    try {
      const res = await fetch(`/api/engagements/${engagementId}/status`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ status })
      });

      if (!res.ok) throw new Error('Failed to update status');
      alert(`Status updated to ${status}!`);
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleRaiseDisputeSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!disputingEngagementId || !disputeReason) return;

    try {
      const res = await fetch(`/api/disputes`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          engagement_id: disputingEngagementId,
          raised_by: user.id,
          reason: disputeReason,
          evidence_urls: disputePhotos ? [disputePhotos] : []
        })
      });

      if (!res.ok) throw new Error('Failed to raise dispute');
      alert('Dispute raised successfully. Worker availability set to dispute hold, admin review in progress.');
      setDisputingEngagementId(null);
      setDisputeReason('');
      setDisputePhotos('');
      fetchEngagements();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleCompleteBooking = async (bookingId: string) => {
    try {
      const res = await fetch(`/api/bookings/${bookingId}/complete`, { method: 'POST' });
      if (!res.ok) throw new Error('Failed to mark booking as completed');
      alert('Booking marked as completed successfully!');
      fetchBookings();
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handleRatingSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!ratingBookingId || !ratingWorkerId) return;

    try {
      const res = await fetch(`/api/bookings/${ratingBookingId}/ratings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          rater_id: user.id,
          ratee_id: ratingWorkerId,
          score: ratingScore,
          comment: ratingComment
        })
      });

      if (!res.ok) throw new Error('Rating submission failed');
      alert('Rating submitted successfully!');
      setRatingBookingId(null);
      setRatingWorkerId(null);
      setRatingComment('');
    } catch (err: any) {
      alert(err.message);
    }
  };

  const handlePaymentSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!paymentBookingId) return;

    try {
      const res = await fetch(`/api/bookings/${paymentBookingId}/payments`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          amount: parseFloat(paymentAmount),
          confirmation_method: paymentMethod
        })
      });

      if (!res.ok) throw new Error('Payment confirmation failed');
      alert('Payment confirmed successfully!');
      setPaymentBookingId(null);
      fetchBookings();
    } catch (err: any) {
      alert(err.message);
    }
  };

  return (
    <div style={{ fontFamily: 'Outfit, sans-serif', backgroundColor: '#f1f5f9', minHeight: '100vh', padding: '24px' }}>
      {/* Header Panel */}
      <header style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', backgroundColor: '#ffffff', padding: '16px 24px', border: '1px solid #e2e8f0', borderRadius: '12px', marginBottom: '24px', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
        <div>
          <h1 style={{ margin: 0, fontSize: '20px', color: '#0f766e', fontWeight: 'bold', letterSpacing: '-0.025em' }}>LabourLink Hiring Portal</h1>
          <p style={{ margin: '4px 0 0 0', fontSize: '13px', color: '#64748b' }}>
            Citizen Client: <strong style={{ color: '#1e293b' }}>{user.full_name}</strong> ({user.phone}) | Role: <strong>Residential Customer</strong>
          </p>
        </div>
        <button 
          onClick={onLogout} 
          style={{ backgroundColor: '#ffffff', color: '#ef4444', border: '1px solid #fecaca', padding: '8px 16px', fontSize: '13px', fontWeight: '600', borderRadius: '8px', cursor: 'pointer', transition: 'all 0.15s ease' }}
        >
          Logout Session
        </button>
      </header>

      {/* Main Grid */}
      <div style={{ display: 'grid', gridTemplateColumns: '2.5fr 1.2fr', gap: '24px' }}>
        
        {/* Left Column */}
        <div>
          {/* Post Service Request */}
          <section className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)', marginBottom: '24px' }}>
            <h3 style={{ margin: '0 0 16px 0', fontSize: '16px', color: '#1e293b', fontWeight: '700' }}>
              CREATE NEW HIRING ENGAGEMENT
            </h3>

            {/* Toggle Hire Worker / Hire Contractor */}
            <div style={{ display: 'flex', gap: '12px', marginBottom: '20px', borderBottom: '1px solid #e2e8f0', paddingBottom: '12px' }}>
              <button 
                type="button" 
                onClick={() => { setHireMode('WORKER'); setRecommendations([]); }}
                style={{
                  flex: 1, padding: '10px', fontSize: '14px', fontWeight: 'bold', borderRadius: '8px', border: 'none', cursor: 'pointer',
                  backgroundColor: hireMode === 'WORKER' ? '#0f766e' : '#f1f5f9',
                  color: hireMode === 'WORKER' ? '#ffffff' : '#475569',
                  transition: 'all 0.2s'
                }}
              >
                💼 Option 1: Hire a Worker
              </button>
              <button 
                type="button" 
                onClick={() => { setHireMode('CONTRACTOR'); setRecommendations([]); }}
                style={{
                  flex: 1, padding: '10px', fontSize: '14px', fontWeight: 'bold', borderRadius: '8px', border: 'none', cursor: 'pointer',
                  backgroundColor: hireMode === 'CONTRACTOR' ? '#0f766e' : '#f1f5f9',
                  color: hireMode === 'CONTRACTOR' ? '#ffffff' : '#475569',
                  transition: 'all 0.2s'
                }}
              >
                🏢 Option 2: Hire a Contractor
              </button>
            </div>
            
            {/* Quick Actions Shortcuts */}
            {hireMode === 'WORKER' && (
              <div style={{ marginBottom: '16px' }}>
                <span style={{ fontSize: '12px', color: '#64748b', display: 'block', marginBottom: '8px', fontWeight: '500' }}>
                  Quick Templates:
                </span>
                <div style={{ display: 'flex', gap: '8px', flexWrap: 'wrap' }}>
                  <button type="button" onClick={() => handleQuickAction('plumber')} style={quickBtnStyle}>🔧 Plumber</button>
                  <button type="button" onClick={() => handleQuickAction('electrician')} style={quickBtnStyle}>⚡ Electrician</button>
                  <button type="button" onClick={() => handleQuickAction('painter')} style={quickBtnStyle}>🎨 Painter</button>
                  <button type="button" onClick={() => handleQuickAction('carpenter')} style={quickBtnStyle}>🪚 Carpenter</button>
                </div>
              </div>
            )}

            <form onSubmit={handleSubmitRequest}>
              <div className="form-group" style={{ marginBottom: '16px' }}>
                <label className="form-label" style={{ fontWeight: '600', color: '#334155', marginBottom: '6px', display: 'block' }}>
                  {hireMode === 'WORKER' ? 'Describe the work needed' : 'Describe the contractor project scope'}
                </label>
                <textarea 
                  className="form-control"
                  style={{ minHeight: '90px', resize: 'vertical', width: '100%', padding: '10px', borderRadius: '8px', border: '1px solid #cbd5e1' }}
                  placeholder={hireMode === 'WORKER' ? "e.g. Need a plumber to fix a leaking tap in the bathroom..." : "e.g. Renovation of a 2BHK flat including tiling, painting, and electric work..."}
                  value={rawText}
                  onChange={(e) => setRawText(e.target.value)}
                  required
                />
              </div>

              {hireMode === 'CONTRACTOR' && (
                <div className="form-group" style={{ marginBottom: '16px' }}>
                  <label className="form-label" style={{ fontWeight: '600', color: '#334155', marginBottom: '6px', display: 'block' }}>
                    Select Contractor
                  </label>
                  <select 
                    value={selectedContractorId}
                    onChange={(e) => setSelectedContractorId(e.target.value)}
                    style={{ width: '100%', padding: '10px', borderRadius: '8px', border: '1px solid #cbd5e1', backgroundColor: '#ffffff' }}
                  >
                    {contractors.map(c => (
                      <option key={c.id} value={c.id}>
                        {c.company_name} ({c.full_name})
                      </option>
                    ))}
                  </select>
                </div>
              )}

              <button 
                type="submit" 
                className="btn btn-primary" 
                style={{ padding: '10px 20px', backgroundColor: '#0f766e', borderColor: '#0f766e', fontWeight: 'bold', width: '100%', borderRadius: '8px', cursor: 'pointer' }}
                disabled={submitting}
              >
                {submitting ? 'Processing request...' : hireMode === 'WORKER' ? 'Search Matches' : 'Book Contractor'}
              </button>
            </form>
          </section>

          {/* Recommendations Feed */}
          {recommendations.length > 0 && (
            <section className="card" style={{ marginBottom: '24px', borderColor: '#99f6e4', backgroundColor: '#f0fdfa', padding: '24px', borderRadius: '12px', border: '1px solid #99f6e4' }}>
              <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '16px' }}>
                <h3 style={{ margin: 0, fontSize: '15px', color: '#0f766e', fontWeight: 'bold' }}>
                  RECOMMENDED WORKERS
                </h3>
                <button 
                  onClick={() => { setRecommendations([]); setActiveReqId(null); }} 
                  style={{ background: 'none', border: 'none', color: '#0d9488', fontSize: '12px', cursor: 'pointer', fontWeight: '600' }}
                >
                  Clear Results
                </button>
              </div>

              <div style={{ display: 'flex', flexDirection: 'column', gap: '16px' }}>
                {recommendations.map((rec) => (
                  <div key={rec.worker_id} style={{ padding: '16px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#ffffff' }}>
                    <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: '8px' }}>
                      <div>
                        <h4 style={{ margin: 0, fontSize: '14px', fontWeight: 'bold', color: '#1e293b' }}>
                          {rec.full_name}
                        </h4>
                        <div style={{ display: 'flex', gap: '4px', flexWrap: 'wrap', marginTop: '4px' }}>
                          {rec.skills.map((s, i) => (
                            <span key={i} style={{ fontSize: '10px', backgroundColor: '#f1f5f9', padding: '2px 6px', borderRadius: '4px', color: '#475569', fontWeight: '500' }}>
                              {s}
                            </span>
                          ))}
                        </div>
                      </div>
                      <div style={{ textAlign: 'right' }}>
                        <div style={{ fontSize: '13px', fontWeight: 'bold', color: '#0f766e' }}>
                          Match: {rec.match_score}%
                        </div>
                        <div style={{ fontSize: '11px', color: '#64748b', marginTop: '2px' }}>
                          Distance: {rec.distance_km} km away
                        </div>
                      </div>
                    </div>

                    <p style={{ margin: '8px 0', fontSize: '12px', color: '#475569', lineHeight: '1.4' }}>
                      {rec.explanation}
                    </p>

                    <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginTop: '12px', paddingTop: '12px', borderTop: '1px solid #f1f5f9' }}>
                      <div style={{ display: 'flex', alignItems: 'center', gap: '8px' }}>
                        <span style={{ fontSize: '12px', color: '#1e293b' }}>
                          Trust Score: <strong>{rec.trust_score !== null ? `${rec.trust_score}/100` : 'Unestablished'}</strong>
                        </span>
                        {rec.trust_score !== null && (
                          <button 
                            onClick={() => setActiveExplainWorkerId(rec.worker_id)} 
                            style={{ background: 'none', border: 'none', color: '#0f766e', fontSize: '11px', textDecoration: 'underline', cursor: 'pointer', padding: 0 }}
                          >
                            Why this score?
                          </button>
                        )}
                      </div>
                      <button 
                        onClick={() => handleBookWorker(rec.worker_id)}
                        style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                      >
                        Book Worker
                      </button>
                    </div>
                  </div>
                ))}
              </div>
            </section>
          )}

          {/* Active Contracts & Negotiations */}
          <section className="card" style={{ marginBottom: '24px', backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
            <h3 style={{ margin: '0 0 16px 0', fontSize: '16px', color: '#1e293b', fontWeight: '700', display: 'flex', alignItems: 'center', gap: '8px' }}>
              🤝 ACTIVE CONTRACTS & NEGOTIATIONS
            </h3>

            {engagements.length === 0 ? (
              <div style={{ padding: '40px', textAlign: 'center', border: '2px dashed #cbd5e1', borderRadius: '8px', color: '#64748b', fontSize: '14px' }}>
                No active service negotiations or engagements found.
              </div>
            ) : (
              <div style={{ display: 'flex', flexDirection: 'column', gap: '16px' }}>
                {engagements.map((eng) => {
                  const lastOffer = eng.offers && eng.offers[0]; // sorted descending by date
                  const isPendingOrNegotiating = eng.status === 'PENDING' || eng.status === 'NEGOTIATING';
                  const lastOfferedByMe = lastOffer && lastOffer.offered_by === user.id;

                  return (
                    <div key={eng.id} style={{ padding: '16px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#f8fafc' }}>
                      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: '8px' }}>
                        <div>
                          <span style={{ fontSize: '10px', color: '#94a3b8', fontWeight: '500' }}>ENGAGEMENT ID: {eng.id.substring(0, 8)}...</span>
                          <h4 style={{ margin: '4px 0 0 0', fontSize: '14px', fontWeight: '700', color: '#1e293b' }}>
                            {eng.mode === 'DIRECT_WORKER' ? '👤 Direct Worker Mode' : '🏢 Via Contractor Mode'} — Counterparty: {eng.counterparty_name}
                          </h4>
                          <p style={{ margin: '6px 0', fontSize: '12px', color: '#475569' }}>
                            Requirement: <strong>"{eng.request_text}"</strong>
                          </p>
                        </div>
                        <span style={{
                          fontSize: '11px',
                          fontWeight: 'bold',
                          padding: '4px 8px',
                          borderRadius: '6px',
                          backgroundColor: eng.status === 'COMPLETED' ? '#e2f5f3' : (eng.status === 'IN_PROGRESS' ? '#dbeafe' : (eng.status === 'DISPUTED' ? '#fee2e2' : '#f1f5f9')),
                          color: eng.status === 'COMPLETED' ? '#0f766e' : (eng.status === 'IN_PROGRESS' ? '#2563eb' : (eng.status === 'DISPUTED' ? '#dc2626' : '#475569'))
                        }}>
                          {eng.status}
                        </span>
                      </div>

                      {/* Offers Log */}
                      {eng.offers && eng.offers.length > 0 && (
                        <div style={{ marginTop: '10px', padding: '10px', backgroundColor: '#ffffff', borderRadius: '6px', border: '1px solid #e2e8f0' }}>
                          <span style={{ fontSize: '11px', fontWeight: 'bold', color: '#64748b', display: 'block', marginBottom: '4px' }}>Negotiation History:</span>
                          <div style={{ display: 'flex', flexDirection: 'column', gap: '4px' }}>
                            {eng.offers.slice(0, 3).map((o: any) => (
                              <div key={o.id} style={{ display: 'flex', justifyContent: 'space-between', fontSize: '12px', color: '#475569' }}>
                                <span>₹{o.amount} proposed by {o.offered_by_name}</span>
                                <span style={{ fontSize: '10px', color: '#94a3b8' }}>{new Date(o.created_at).toLocaleTimeString()}</span>
                              </div>
                            ))}
                          </div>
                        </div>
                      )}

                      {/* Control buttons & Counter Offers */}
                      <div style={{ marginTop: '12px', display: 'flex', flexWrap: 'wrap', gap: '8px', alignItems: 'center' }}>
                        {isPendingOrNegotiating && (
                          <>
                            {!lastOfferedByMe ? (
                              <>
                                <button
                                  onClick={() => handleRespondToProposal(eng.id, 'ACCEPTED')}
                                  style={{ backgroundColor: '#10b981', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                >
                                  Accept Offer (₹{lastOffer?.amount || 500})
                                </button>
                                <button
                                  onClick={() => handleRespondToProposal(eng.id, 'REJECTED')}
                                  style={{ backgroundColor: '#ef4444', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                                >
                                  Reject Offer
                                </button>
                              </>
                            ) : (
                              <span style={{ fontSize: '12px', color: '#64748b', fontStyle: 'italic' }}>Awaiting counterparty response...</span>
                            )}
                            
                            <div style={{ display: 'flex', gap: '4px', marginLeft: 'auto' }}>
                              <input
                                type="number"
                                placeholder="Counter offer price"
                                value={counterProposalAmount[eng.id] || ''}
                                onChange={(e) => setCounterProposalAmount({ ...counterProposalAmount, [eng.id]: e.target.value })}
                                style={{ width: '130px', padding: '5px 8px', fontSize: '12px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                              />
                              <button
                                onClick={() => handleProposePriceOffer(eng.id)}
                                style={{ backgroundColor: '#0f766e', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                              >
                                Counter
                              </button>
                            </div>
                          </>
                        )}

                        {eng.status === 'ACCEPTED' && (
                          <>
                            <button
                              onClick={() => handleUpdateEngagementStatus(eng.id, 'IN_PROGRESS')}
                              style={{ backgroundColor: '#3b82f6', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                            >
                              Start Work
                            </button>
                            <button
                              onClick={() => handleUpdateEngagementStatus(eng.id, 'CANCELLED')}
                              style={{ backgroundColor: '#ef4444', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                            >
                              Cancel Engagement
                            </button>
                          </>
                        )}

                        {eng.status === 'IN_PROGRESS' && (
                          <>
                            <button
                              onClick={() => handleUpdateEngagementStatus(eng.id, 'COMPLETED')}
                              style={{ backgroundColor: '#10b981', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                            >
                              Mark Completed
                            </button>
                            <button
                              onClick={() => setDisputingEngagementId(eng.id)}
                              style={{ backgroundColor: '#ef4444', color: '#ffffff', border: 'none', padding: '6px 12px', fontSize: '12px', fontWeight: 'bold', borderRadius: '6px', cursor: 'pointer' }}
                            >
                              Raise Dispute
                            </button>
                          </>
                        )}

                        {eng.status === 'DISPUTED' && (
                          <div style={{ color: '#dc2626', fontSize: '12px', fontWeight: '600' }}>
                            ⚠️ Dispute raised. Worker status set to busy. Admin arbiters reviewing case.
                          </div>
                        )}
                      </div>
                    </div>
                  );
                })}
              </div>
            )}
          </section>

          {/* Bookings List */}
          <section className="card" style={{ backgroundColor: '#ffffff', padding: '24px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
            <h3 style={{ margin: '0 0 16px 0', fontSize: '16px', color: '#1e293b', fontWeight: '700' }}>
              ENGAGEMENT HISTORY
            </h3>

            {bookings.length === 0 ? (
              <div style={{ padding: '40px', textAlign: 'center', border: '2px dashed #cbd5e1', borderRadius: '8px', color: '#64748b', fontSize: '14px' }}>
                No active bookings or contractor projects found.
              </div>
            ) : (
              <div style={{ display: 'flex', flexDirection: 'column', gap: '16px' }}>
                {bookings.map((b) => (
                  <div key={b.id} style={{ padding: '16px', border: '1px solid #e2e8f0', borderRadius: '8px', backgroundColor: '#ffffff' }}>
                    <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start', marginBottom: '8px' }}>
                      <div>
                        <span style={{ fontSize: '10px', color: '#94a3b8', fontWeight: '500' }}>ID: {b.id.substring(0, 8)}...</span>
                        <h4 style={{ margin: '4px 0 0 0', fontSize: '15px', fontWeight: '700', color: '#1e293b' }}>
                          {b.contractor_id ? `🏢 Contractor: ${b.contractor_name} (${b.contractor_company})` : `👤 Worker: ${b.worker_name}`}
                        </h4>
                      </div>
                      <span style={{ 
                        fontSize: '11px', 
                        fontWeight: 'bold', 
                        padding: '4px 8px', 
                        borderRadius: '6px', 
                        backgroundColor: b.status === 'COMPLETED' ? '#e2f5f3' : (b.status === 'ACCEPTED' || b.status === 'IN_PROGRESS' ? '#fffbeb' : '#f1f5f9'), 
                        color: b.status === 'COMPLETED' ? '#0f766e' : (b.status === 'ACCEPTED' || b.status === 'IN_PROGRESS' ? '#d97706' : '#475569') 
                      }}>
                        {b.status}
                      </span>
                    </div>

                    <p style={{ margin: '8px 0', fontSize: '13px', color: '#475569' }}>
                      Requirement: <em>"{b.requirement_text}"</em>
                    </p>

                    {/* Contractor Assignment Display */}
                    {b.contractor_id && bookingAssignments[b.id] && bookingAssignments[b.id].length > 0 && (
                      <div style={{ marginTop: '12px', padding: '12px', backgroundColor: '#f8fafc', borderRadius: '8px', border: '1px solid #e2e8f0' }}>
                        <span style={{ fontSize: '11px', fontWeight: '700', color: '#475569', display: 'block', marginBottom: '6px' }}>
                          Contractor Worker Assignments:
                        </span>
                        <div style={{ display: 'flex', flexDirection: 'column', gap: '6px' }}>
                          {bookingAssignments[b.id].map((asn: any) => (
                            <div key={asn.id} style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', fontSize: '12px' }}>
                              <span style={{ fontWeight: '500', color: '#1e293b' }}>👤 {asn.worker_name}</span>
                              <span style={{ 
                                fontSize: '10px', 
                                fontWeight: '700',
                                padding: '2px 6px',
                                borderRadius: '4px',
                                backgroundColor: asn.status === 'ACCEPTED' ? '#d1fae5' : asn.status === 'ASSIGNED' ? '#fef3c7' : '#f1f5f9',
                                color: asn.status === 'ACCEPTED' ? '#065f46' : asn.status === 'ASSIGNED' ? '#92400e' : '#475569'
                              }}>{asn.status}</span>
                            </div>
                          ))}
                        </div>
                      </div>
                    )}

                    {b.payment_amount && (
                      <div style={{ fontSize: '12px', color: '#1e3a8a', backgroundColor: '#eff6ff', padding: '6px 12px', borderRadius: '6px', marginTop: '8px', fontWeight: '500' }}>
                        Payment Verified: <strong>Rs. {b.payment_amount}</strong> ({b.payment_status})
                      </div>
                    )}

                    {/* Booking Actions */}
                    <div style={{ display: 'flex', gap: '8px', marginTop: '12px', paddingTop: '12px', borderTop: '1px solid #f1f5f9', justifyContent: 'flex-end' }}>
                      {(b.status === 'ACCEPTED' || b.status === 'IN_PROGRESS') && (
                        <button 
                          onClick={() => handleCompleteBooking(b.id)}
                          style={actionBtnStyle}
                        >
                          Complete Booking
                        </button>
                      )}
                      
                      {b.status === 'COMPLETED' && !b.payment_amount && (
                        <button 
                          onClick={() => { setPaymentBookingId(b.id); setPaymentAmount('500'); }}
                          style={{ ...actionBtnStyle, backgroundColor: '#d97706', color: '#ffffff' }}
                        >
                          💸 Confirm Payment
                        </button>
                      )}

                      {b.status === 'COMPLETED' && (
                        <button 
                          onClick={() => { setRatingBookingId(b.id); setRatingWorkerId(b.worker_id || ''); }}
                          style={actionBtnStyle}
                          disabled={!b.worker_id}
                        >
                          ⭐ Rate Worker
                        </button>
                      )}
                    </div>
                  </div>
                ))}
              </div>
            )}
          </section>
        </div>

        {/* Right Column: Timeline */}
        <div>
          <section className="card" style={{ backgroundColor: '#ffffff', padding: '20px', borderRadius: '12px', border: '1px solid #e2e8f0', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.05)' }}>
            <ActivityTimeline userId={user.id} />
          </section>
        </div>
      </div>

      {/* Trust explanation modal */}
      {activeExplainWorkerId && (
        <TrustScoreBreakdownModal 
          workerId={activeExplainWorkerId} 
          onClose={() => setActiveExplainWorkerId(null)} 
        />
      )}

      {/* Rate worker modal */}
      {ratingBookingId && (
        <div style={backdropStyle}>
          <div style={modalStyle}>
            <div style={{ padding: '20px', borderBottom: '1px solid #cbd5e1' }}>
              <h3 style={{ margin: 0, fontSize: '15px', color: '#1e293b', fontWeight: 'bold' }}>Rate Worker Performance</h3>
            </div>
            <form onSubmit={handleRatingSubmit}>
              <div style={{ padding: '20px' }}>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label">Score (1 to 5 Stars)</label>
                  <select 
                    className="form-control"
                    value={ratingScore} 
                    onChange={(e) => setRatingScore(parseInt(e.target.value))}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                  >
                    <option value={5}>5 Stars - Excellent (बहुत बढ़िया)</option>
                    <option value={4}>4 Stars - Good (अच्छा)</option>
                    <option value={3}>3 Stars - Average (ठीक ठाक)</option>
                    <option value={2}>2 Stars - Poor (खराब)</option>
                    <option value={1}>1 Star - Critical / Collusion Test (बहुत खराब)</option>
                  </select>
                </div>
                <div className="form-group">
                  <label className="form-label">Comments / Review</label>
                  <textarea 
                    className="form-control"
                    value={ratingComment}
                    onChange={(e) => setRatingComment(e.target.value)}
                    placeholder="Enter honest feedback. Rating cluster patterns are audited by Trust Agent."
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1', minHeight: '60px' }}
                    required
                  />
                </div>
              </div>
              <div style={modalFooterStyle}>
                <button type="button" onClick={() => { setRatingBookingId(null); setRatingWorkerId(null); }} style={cancelBtnStyle}>Cancel</button>
                <button type="submit" className="btn btn-primary" style={{ padding: '6px 12px', backgroundColor: '#0f766e', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer' }}>Submit Rating</button>
              </div>
            </form>
          </div>
        </div>
      )}

      {/* Payment confirmation modal */}
      {paymentBookingId && (
        <div style={backdropStyle}>
          <div style={modalStyle}>
            <div style={{ padding: '20px', borderBottom: '1px solid #cbd5e1' }}>
              <h3 style={{ margin: 0, fontSize: '15px', color: '#1e293b', fontWeight: 'bold' }}>Attest & Confirm Payment</h3>
            </div>
            <form onSubmit={handlePaymentSubmit}>
              <div style={{ padding: '20px' }}>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label">UPI Amount Paid (Rs.)</label>
                  <input 
                    type="number"
                    className="form-control"
                    value={paymentAmount}
                    onChange={(e) => setPaymentAmount(e.target.value)}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                    required
                  />
                </div>
                <div className="form-group">
                  <label className="form-label">Attestation Channel</label>
                  <select 
                    className="form-control"
                    value={paymentMethod}
                    onChange={(e) => setPaymentMethod(e.target.value as any)}
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                  >
                    <option value="UPI_VERIFIED">UPI Transaction Verification</option>
                    <option value="CASH_ATTESTED">Cash Hand-to-Hand Attested</option>
                    <option value="BANK_VERIFIED">Direct IMPS/Bank Transferred</option>
                  </select>
                </div>
              </div>
              <div style={modalFooterStyle}>
                <button type="button" onClick={() => setPaymentBookingId(null)} style={cancelBtnStyle}>Cancel</button>
                <button type="submit" className="btn btn-primary" style={{ padding: '6px 12px', backgroundColor: '#0f766e', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer' }}>Attest Payment</button>
              </div>
            </form>
          </div>
        </div>
      )}

      {/* Raise Dispute Modal */}
      {disputingEngagementId && (
        <div style={backdropStyle}>
          <div style={modalStyle}>
            <div style={{ padding: '20px', borderBottom: '1px solid #cbd5e1' }}>
              <h3 style={{ margin: 0, fontSize: '15px', color: '#dc2626', fontWeight: 'bold' }}>Raise Dispute / Conflict Audit</h3>
            </div>
            <form onSubmit={handleRaiseDisputeSubmit}>
              <div style={{ padding: '20px' }}>
                <div className="form-group" style={{ marginBottom: '12px' }}>
                  <label className="form-label" style={{ color: '#475569', fontWeight: '600' }}>Reason for Dispute</label>
                  <textarea 
                    className="form-control"
                    value={disputeReason}
                    onChange={(e) => setDisputeReason(e.target.value)}
                    placeholder="Describe the issue in detail. The dispute will be escalated to Admin arbiters..."
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1', minHeight: '80px' }}
                    required
                  />
                </div>
                <div className="form-group">
                  <label className="form-label" style={{ color: '#475569', fontWeight: '600' }}>Evidence Photo URL (Optional)</label>
                  <input 
                    type="text"
                    className="form-control"
                    value={disputePhotos}
                    onChange={(e) => setDisputePhotos(e.target.value)}
                    placeholder="https://evidence.domain.com/photo.jpg"
                    style={{ width: '100%', padding: '8px', borderRadius: '6px', border: '1px solid #cbd5e1' }}
                  />
                </div>
              </div>
              <div style={modalFooterStyle}>
                <button type="button" onClick={() => setDisputingEngagementId(null)} style={cancelBtnStyle}>Cancel</button>
                <button type="submit" className="btn btn-primary" style={{ padding: '6px 12px', backgroundColor: '#dc2626', border: 'none', color: '#ffffff', borderRadius: '6px', cursor: 'pointer' }}>Submit Dispute</button>
              </div>
            </form>
          </div>
        </div>
      )}

    </div>
  );
};

const quickBtnStyle: React.CSSProperties = {
  backgroundColor: '#f8fafc',
  border: '1px solid #cbd5e1',
  padding: '6px 12px',
  borderRadius: '6px',
  fontSize: '12px',
  cursor: 'pointer',
  color: '#334155',
  fontWeight: '500',
  transition: 'all 0.15s ease'
};

const actionBtnStyle: React.CSSProperties = {
  backgroundColor: '#f1f5f9',
  border: '1px solid #cbd5e1',
  color: '#334155',
  padding: '6px 12px',
  fontSize: '12px',
  fontWeight: 'bold',
  borderRadius: '6px',
  cursor: 'pointer'
};

const backdropStyle: React.CSSProperties = {
  position: 'fixed',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  backgroundColor: 'rgba(15, 23, 42, 0.6)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  zIndex: 999
};

const modalStyle: React.CSSProperties = {
  backgroundColor: '#ffffff',
  width: '100%',
  maxWidth: '440px',
  borderRadius: '8px',
  boxShadow: '0 10px 15px -3px rgb(0 0 0 / 0.1)',
  overflow: 'hidden'
};

const modalFooterStyle: React.CSSProperties = {
  padding: '12px 20px',
  backgroundColor: '#f8fafc',
  borderTop: '1px solid #cbd5e1',
  display: 'flex',
  justifyContent: 'flex-end',
  gap: '8px'
};

const cancelBtnStyle: React.CSSProperties = {
  background: 'none',
  border: '1px solid #cbd5e1',
  padding: '6px 12px',
  borderRadius: '6px',
  fontSize: '13px',
  cursor: 'pointer',
  color: '#475569'
};

export default CustomerDashboard;
